{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "High-resolution profile of neoantigen-specific TCR activation links moderate stimulation to resilience and sustained tumor control\n",
    "\n",
    "Franziska Füchsl*, Johannes Untch*, Vladyslav Kavaka, Sebastian Jarosch, Carolin Vogelsang, Niklas de Andrade Krätzig, Dario Gosmann, Roland Rad, Dirk Busch, Eduardo Beltrán, Eva Bräunlein#, Angela M. Krackhardt#\n",
    "\n",
    "Analysis by: Vladyslav Kavaka (vladyslav.kavaka@med.uni-muenchen.de), Supervision: Eduardo Beltrán (eduardo.beltran@med.uni-muenchen.de)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.0.5 (2021-03-31)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 18.04.6 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /home/INIM/vladyslav.kavaka/miniconda3/envs/azimuth/lib/libopenblasp-r0.3.17.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=C.UTF-8    LC_NUMERIC=C        LC_TIME=C          \n",
       " [4] LC_COLLATE=C        LC_MONETARY=C       LC_MESSAGES=C      \n",
       " [7] LC_PAPER=C          LC_NAME=C           LC_ADDRESS=C       \n",
       "[10] LC_TELEPHONE=C      LC_MEASUREMENT=C    LC_IDENTIFICATION=C\n",
       "\n",
       "attached base packages:\n",
       "[1] parallel  stats4    stats     graphics  grDevices utils     datasets \n",
       "[8] methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] RColorBrewer_1.1-3             SeuratDisk_0.0.0.9019         \n",
       " [3] bonemarrowref.SeuratData_1.0.0 SeuratData_0.2.2              \n",
       " [5] ggupset_0.3.0                  ggrastr_1.0.1                 \n",
       " [7] celldex_1.0.0                  Nebulosa_1.0.2                \n",
       " [9] harmony_0.1.0                  Rcpp_1.0.8                    \n",
       "[11] enrichR_3.0                    qpcR_1.4-1                    \n",
       "[13] robustbase_0.93-9              rgl_0.108.3                   \n",
       "[15] minpack.lm_1.2-1               MASS_7.3-55                   \n",
       "[17] MAST_1.16.0                    SingleCellExperiment_1.12.0   \n",
       "[19] SummarizedExperiment_1.20.0    Biobase_2.50.0                \n",
       "[21] GenomicRanges_1.42.0           GenomeInfoDb_1.26.4           \n",
       "[23] IRanges_2.24.1                 S4Vectors_0.28.1              \n",
       "[25] BiocGenerics_0.36.0            MatrixGenerics_1.2.1          \n",
       "[27] matrixStats_0.61.0             purrr_0.3.4                   \n",
       "[29] scales_1.2.1                   ggrepel_0.9.1                 \n",
       "[31] gprofiler2_0.2.1               patchwork_1.1.2               \n",
       "[33] ggthemes_4.2.4                 ggplot2_3.4.0                 \n",
       "[35] limma_3.46.0                   tidyr_1.1.4                   \n",
       "[37] Matrix_1.4-0                   dplyr_1.0.7                   \n",
       "[39] SeuratObject_4.0.4             Seurat_4.1.0                  \n",
       "[41] vctrs_0.5.1                    devtools_2.4.5                \n",
       "[43] usethis_2.1.6                 \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] utf8_1.2.2                    reticulate_1.23              \n",
       "  [3] ks_1.13.5                     tidyselect_1.1.1             \n",
       "  [5] RSQLite_2.2.9                 AnnotationDbi_1.52.0         \n",
       "  [7] htmlwidgets_1.5.4             grid_4.0.5                   \n",
       "  [9] Rtsne_0.15                    munsell_0.5.0                \n",
       " [11] codetools_0.2-18              ica_1.0-2                    \n",
       " [13] pbdZMQ_0.3-6                  future_1.23.0                \n",
       " [15] miniUI_0.1.1.1                withr_2.5.0                  \n",
       " [17] colorspace_2.0-3              knitr_1.37                   \n",
       " [19] uuid_1.0-3                    ROCR_1.0-11                  \n",
       " [21] tensor_1.5                    listenv_0.8.0                \n",
       " [23] repr_1.1.4                    GenomeInfoDbData_1.2.4       \n",
       " [25] polyclip_1.10-0               bit64_4.0.5                  \n",
       " [27] parallelly_1.30.0             generics_0.1.1               \n",
       " [29] xfun_0.29                     BiocFileCache_1.14.0         \n",
       " [31] R6_2.5.1                      ggbeeswarm_0.6.0             \n",
       " [33] hdf5r_1.3.5                   bitops_1.0-7                 \n",
       " [35] spatstat.utils_2.3-0          cachem_1.0.6                 \n",
       " [37] DelayedArray_0.16.3           assertthat_0.2.1             \n",
       " [39] promises_1.2.0.1              beeswarm_0.4.0               \n",
       " [41] gtable_0.3.1                  globals_0.14.0               \n",
       " [43] processx_3.5.2                goftest_1.2-3                \n",
       " [45] rlang_1.0.6                   splines_4.0.5                \n",
       " [47] lazyeval_0.2.2                spatstat.geom_2.4-0          \n",
       " [49] yaml_2.2.1                    BiocManager_1.30.16          \n",
       " [51] reshape2_1.4.4                abind_1.4-5                  \n",
       " [53] httpuv_1.6.5                  tools_4.0.5                  \n",
       " [55] ellipsis_0.3.2                spatstat.core_2.3-2          \n",
       " [57] sessioninfo_1.2.2             ggridges_0.5.3               \n",
       " [59] plyr_1.8.6                    sparseMatrixStats_1.2.1      \n",
       " [61] base64enc_0.1-3               zlibbioc_1.36.0              \n",
       " [63] RCurl_1.98-1.6                ps_1.6.0                     \n",
       " [65] prettyunits_1.1.1             rpart_4.1.16                 \n",
       " [67] deldir_1.0-6                  pbapply_1.5-0                \n",
       " [69] cowplot_1.1.1                 urlchecker_1.0.1             \n",
       " [71] zoo_1.8-9                     cluster_2.1.2                \n",
       " [73] fs_1.5.2                      magrittr_2.0.3               \n",
       " [75] data.table_1.14.2             scattermore_0.7              \n",
       " [77] lmtest_0.9-39                 RANN_2.6.1                   \n",
       " [79] mvtnorm_1.1-3                 fitdistrplus_1.1-6           \n",
       " [81] pkgload_1.3.2                 mime_0.12                    \n",
       " [83] evaluate_0.14                 xtable_1.8-4                 \n",
       " [85] mclust_5.4.9                  gridExtra_2.3                \n",
       " [87] compiler_4.0.5                tibble_3.1.8                 \n",
       " [89] KernSmooth_2.23-20            crayon_1.4.2                 \n",
       " [91] htmltools_0.5.2               mgcv_1.8-38                  \n",
       " [93] later_1.2.0                   DBI_1.1.2                    \n",
       " [95] ExperimentHub_1.16.1          dbplyr_2.1.1                 \n",
       " [97] rappdirs_0.3.3                cli_3.4.1                    \n",
       " [99] igraph_1.2.11                 pkgconfig_2.0.3              \n",
       "[101] pkgdown_2.0.6                 IRdisplay_1.1                \n",
       "[103] plotly_4.10.0                 spatstat.sparse_2.1-0        \n",
       "[105] vipor_0.4.5                   XVector_0.30.0               \n",
       "[107] stringr_1.4.0                 callr_3.7.0                  \n",
       "[109] digest_0.6.29                 sctransform_0.3.3            \n",
       "[111] RcppAnnoy_0.0.19              pracma_2.3.8                 \n",
       "[113] spatstat.data_2.1-2           leiden_0.3.9                 \n",
       "[115] uwot_0.1.11                   DelayedMatrixStats_1.12.3    \n",
       "[117] curl_4.3.2                    shiny_1.7.1                  \n",
       "[119] rjson_0.2.21                  lifecycle_1.0.3              \n",
       "[121] nlme_3.1-152                  jsonlite_1.7.2               \n",
       "[123] viridisLite_0.4.1             fansi_1.0.3                  \n",
       "[125] pillar_1.8.1                  lattice_0.20-45              \n",
       "[127] fastmap_1.1.0                 httr_1.4.2                   \n",
       "[129] DEoptimR_1.0-10               pkgbuild_1.3.1               \n",
       "[131] survival_3.2-13               interactiveDisplayBase_1.28.0\n",
       "[133] glue_1.6.2                    remotes_2.4.2                \n",
       "[135] png_0.1-7                     BiocVersion_3.12.0           \n",
       "[137] bit_4.0.4                     stringi_1.7.6                \n",
       "[139] profvis_0.3.7                 blob_1.2.2                   \n",
       "[141] AnnotationHub_2.22.1          memoise_2.0.1                \n",
       "[143] IRkernel_1.3                  irlba_2.3.5                  \n",
       "[145] future.apply_1.8.1           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "'/home/INIM/vladyslav.kavaka/miniconda3/envs/azimuth/lib/R/library'"
      ],
      "text/latex": [
       "'/home/INIM/vladyslav.kavaka/miniconda3/envs/azimuth/lib/R/library'"
      ],
      "text/markdown": [
       "'/home/INIM/vladyslav.kavaka/miniconda3/envs/azimuth/lib/R/library'"
      ],
      "text/plain": [
       "[1] \"/home/INIM/vladyslav.kavaka/miniconda3/envs/azimuth/lib/R/library\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#For the R part of the code\n",
    "sessionInfo()\n",
    "set.seed(1)\n",
    ".libPaths()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "matplotlib          3.6.3\n",
       "numpy               1.22.3\n",
       "pandas              1.4.2\n",
       "scanpy              1.9.1\n",
       "scipy               1.10.0\n",
       "session_info        1.0.0\n",
       "umap                0.5.3\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "PIL                                         8.4.0\n",
       "aa8f2297d25b4dc6fd3d98411eb3ba53823c4f42    NA\n",
       "absl                                        NA\n",
       "anndata                                     0.8.0\n",
       "asttokens                                   NA\n",
       "astunparse                                  1.6.3\n",
       "backcall                                    0.2.0\n",
       "brotli                                      NA\n",
       "certifi                                     2022.12.07\n",
       "cffi                                        1.15.0\n",
       "charset_normalizer                          2.0.10\n",
       "cloudpickle                                 2.2.1\n",
       "colorama                                    0.4.6\n",
       "cycler                                      0.10.0\n",
       "cython_runtime                              NA\n",
       "dateutil                                    2.8.2\n",
       "debugpy                                     1.5.1\n",
       "decorator                                   5.1.1\n",
       "defusedxml                                  0.7.1\n",
       "entrypoints                                 0.3\n",
       "executing                                   0.8.2\n",
       "flatbuffers                                 23.1.21\n",
       "gast                                        NA\n",
       "google                                      NA\n",
       "h5py                                        3.6.0\n",
       "idna                                        3.3\n",
       "igraph                                      0.9.9\n",
       "ipykernel                                   6.15.0\n",
       "ipython_genutils                            0.2.0\n",
       "jedi                                        0.18.1\n",
       "joblib                                      1.2.0\n",
       "jupyter_server                              1.13.3\n",
       "keras                                       2.11.0\n",
       "kiwisolver                                  1.4.2\n",
       "leidenalg                                   0.8.9\n",
       "llvmlite                                    0.34.0\n",
       "mpl_toolkits                                NA\n",
       "natsort                                     8.2.0\n",
       "numba                                       0.51.2\n",
       "opt_einsum                                  v3.3.0\n",
       "packaging                                   21.3\n",
       "parso                                       0.8.3\n",
       "pexpect                                     4.8.0\n",
       "pickleshare                                 0.7.5\n",
       "pkg_resources                               NA\n",
       "prompt_toolkit                              3.0.24\n",
       "psutil                                      5.9.0\n",
       "ptyprocess                                  0.7.0\n",
       "pure_eval                                   0.2.1\n",
       "pycparser                                   2.21\n",
       "pydev_ipython                               NA\n",
       "pydevconsole                                NA\n",
       "pydevd                                      2.6.0\n",
       "pydevd_concurrency_analyser                 NA\n",
       "pydevd_file_utils                           NA\n",
       "pydevd_plugins                              NA\n",
       "pydevd_tracing                              NA\n",
       "pygments                                    2.11.2\n",
       "pynndescent                                 0.5.8\n",
       "pyparsing                                   3.0.6\n",
       "pytz                                        2021.3\n",
       "requests                                    2.27.1\n",
       "setuptools                                  60.5.0\n",
       "six                                         1.16.0\n",
       "sklearn                                     0.24.2\n",
       "socks                                       1.7.1\n",
       "stack_data                                  0.1.4\n",
       "tensorboard                                 2.11.2\n",
       "tensorflow                                  2.11.0\n",
       "tensorflow_probability                      0.19.0\n",
       "termcolor                                   NA\n",
       "texttable                                   1.6.7\n",
       "tornado                                     6.1\n",
       "tqdm                                        4.64.1\n",
       "traitlets                                   5.1.1\n",
       "tree                                        0.1.8\n",
       "typing_extensions                           NA\n",
       "urllib3                                     1.26.8\n",
       "wcwidth                                     0.2.5\n",
       "wrapt                                       1.14.1\n",
       "zmq                                         22.3.0\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             8.0.0\n",
       "jupyter_client      7.1.1\n",
       "jupyter_core        4.9.1\n",
       "jupyterlab          3.2.8\n",
       "notebook            6.4.7\n",
       "-----\n",
       "Python 3.8.12 | packaged by conda-forge | (default, Oct 12 2021, 21:59:51) [GCC 9.4.0]\n",
       "Linux-5.4.114-1-pve-x86_64-with-glibc2.10\n",
       "-----\n",
       "Session information updated at 2023-01-30 17:11\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#For the python based part\n",
    "session_info.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(devtools)\n",
    "library(vctrs)\n",
    "library(Seurat)\n",
    "library(dplyr)\n",
    "library(Matrix)\n",
    "library(tidyr)\n",
    "library(limma)\n",
    "library(ggplot2)\n",
    "library(ggthemes)\n",
    "library(patchwork)\n",
    "library(gprofiler2)\n",
    "library(ggrepel)\n",
    "library(scales)\n",
    "library(ggthemes)\n",
    "library(purrr)\n",
    "library(MAST)\n",
    "library(qpcR)\n",
    "library(enrichR)\n",
    "library(harmony)\n",
    "library(Nebulosa)\n",
    "library(celldex)\n",
    "library(ggrastr)\n",
    "library(ggupset)\n",
    "library(SeuratData)\n",
    "library(SeuratDisk)\n",
    "library(RColorBrewer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=11, repr.plot.height=11)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Read 10X data for the first sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read 10X data:\n",
    "matrix_dir = \"~/pathway_to_data/outs/filtered_feature_bc_matrix/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "list.files(matrix_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- Read10X(data.dir = matrix_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "str(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create Seurat object\n",
    "pbmc <- CreateSeuratObject (counts = data, min.cells = 3, min.features = 200, project = \"pbmc_1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The number of features and UMIs (nFeature_RNA and nCount_RNA) are automatically calculated for every object by Seurat.\n",
    "# For non-UMI data, nCount_RNA represents the sum of the non-normalized values within a cell\n",
    "# We calculate the percentage of mitochondrial features here and store it in object metadata as `percent.mito`.\n",
    "# We use raw count data since this represents non-transformed and non-log-normalized counts\n",
    "# The % of UMI mapping to MT-features is a common scRNA-seq QC metric.\n",
    "mito.features <- grep(pattern = \"^MT-\", x = rownames(x = pbmc), value = TRUE)\n",
    "percent.mito <- Matrix::colSums(x = GetAssayData(object = pbmc, slot = 'counts')[mito.features, ]) / Matrix::colSums(x = GetAssayData(object = pbmc, slot = 'counts'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# The [[ operator can add columns to object metadata, and is a great place to stash QC stats\n",
    "pbmc[['percent.mito']] <- percent.mito\n",
    "VlnPlot(object = pbmc, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mito\"), ncol = 3, pt.size = 0.000001)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FeatureScatter is typically used to visualize feature-feature relationships, but can be used for anything \n",
    "# calculated by the object, i.e. columns in object metadata, PC scores etc.\n",
    "# Since there is a rare subset of cells with an outlier level of high mitochondrial percentage\n",
    "# and also low UMI content, we filter these as well\n",
    "FeatureScatter(object = pbmc, feature1 = \"nCount_RNA\", feature2 = \"percent.mito\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FeatureScatter(object = pbmc, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We filter out cells that have unique feature counts over 5,000 or less than 200\n",
    "pbmc <- subset(x = pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & percent.mito < '0.18')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add TCR information of the first sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#First we need to create TCR file for the first sample in the list\n",
    "tcr <- read.csv(paste('~/pathway_to_data/cellranger_vdj/outs/G1=stimulated/outs/', \"filtered_contig_annotations.csv\", sep=\"\"))\n",
    "tcr <- with(tcr, tcr[order(chain, decreasing = TRUE), ]) # place TRB on top before removing duplicates\n",
    "tcr <- tcr[!duplicated(tcr$barcode), ]\n",
    "#choose the columns to keep\n",
    "tcr <- tcr[,c(\"barcode\", \"raw_clonotype_id\", \"chain\", 'v_gene')]\n",
    "names(tcr)[names(tcr) == \"raw_clonotype_id\"] <- \"clonotype_id\"\n",
    "#read clonotypes file\n",
    "clono <- read.csv(paste('~/pathway_to_TCR_data/cellranger_vdj/outs/G1=stimulated/outs/', \"clonotypes.csv\", sep=\"\"))\n",
    "tcr <- merge(tcr, clono[, c(\"clonotype_id\", \"frequency\", \"cdr3s_aa\")])\n",
    "#Rename columns\n",
    "names(tcr)[1] <- \"TCR_clonotype_id\"\n",
    "names(tcr)[3] <- 'TCR_chain'\n",
    "names(tcr)[4] <- 'TCR_v_gene'\n",
    "names(tcr)[5] <- 'TCR_frequency'\n",
    "names(tcr)[6] <- 'TCR_cdr3'\n",
    "#reorder Columns\n",
    "tcr <- tcr[, c(2, 1, 3, 4, 5, 6)]\n",
    "#correct rownames\n",
    "rownames(tcr) <- tcr[,1]\n",
    "tcr[,1] <- NULL\n",
    "#Split cdr3 column:\n",
    "tcr <- separate(data = tcr, col = TCR_cdr3, into = c(\"TCR1\", \"TCR2\", \"TCR3\", \"TCR4\"), sep = \"\\\\;\")\n",
    "tcr[is.na(tcr)] <- \"FALSE\"\n",
    "head(tcr)\n",
    "tcr.combined <- tcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#divide in TRA and TRB subset:\n",
    "for (k in 1:nrow(tcr.combined)){\n",
    "  if(startsWith(tcr.combined$TCR1[k], 'TRB:')){\n",
    "    tcr.combined$TCR1B[k] <- sub(pattern = '.*:', x = tcr.combined$TCR1[k], '')\n",
    "  } else {tcr.combined$TCR1B[k] <- 'FALSE'}\n",
    "    if(startsWith(tcr.combined$TCR1[k], 'TRA:')){\n",
    "    tcr.combined$TCR1A[k] <- sub(pattern = '.*:', x = tcr.combined$TCR1[k], '')\n",
    "  } else {tcr.combined$TCR1A[k] <- 'FALSE'}\n",
    "  \n",
    "  if(startsWith(tcr.combined$TCR2[k], 'TRB:')){\n",
    "    tcr.combined$TCR2B[k] <- sub(pattern = '.*:', x = tcr.combined$TCR2[k], '')\n",
    "  } else {tcr.combined$TCR2B[k] <- 'FALSE'}\n",
    "        if(startsWith(tcr.combined$TCR2[k], 'TRA:')){\n",
    "    tcr.combined$TCR2A[k] <- sub(pattern = '.*:', x = tcr.combined$TCR2[k], '')\n",
    "  } else {tcr.combined$TCR2A[k] <- 'FALSE'}\n",
    "  \n",
    "  if(startsWith(tcr.combined$TCR3[k], 'TRB:')){\n",
    "    tcr.combined$TCR3B[k] <- sub(pattern = '.*:', x = tcr.combined$TCR3[k], '')\n",
    "  } else {tcr.combined$TCR3B[k] <- 'FALSE'}\n",
    "     if(startsWith(tcr.combined$TCR3[k], 'TRA:')){\n",
    "    tcr.combined$TCR3A[k] <- sub(pattern = '.*:', x = tcr.combined$TCR3[k], '')\n",
    "  } else {tcr.combined$TCR3A[k] <- 'FALSE'}\n",
    "  \n",
    "  if(startsWith(tcr.combined$TCR4[k], 'TRB:')){\n",
    "    tcr.combined$TCR4B[k] <- sub(pattern = '.*:', x = tcr.combined$TCR4[k], '')\n",
    "  } else {tcr.combined$TCR4B[k] <- 'FALSE'}\n",
    "    if(startsWith(tcr.combined$TCR4[k], 'TRA:')){\n",
    "    tcr.combined$TCR4A[k] <- sub(pattern = '.*:', x = tcr.combined$TCR4[k], '')\n",
    "  } else {tcr.combined$TCR4A[k] <- 'FALSE'}\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcr.combined$TCR1 <- NULL\n",
    "tcr.combined$TCR2 <- NULL\n",
    "tcr.combined$TCR3 <- NULL\n",
    "tcr.combined$TCR4 <- NULL\n",
    "head(tcr.combined)\n",
    "tail(tcr.combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nrow(tcr.combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(tcr.combined, file = './tcr_1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc <- AddMetaData(object = pbmc, metadata = tcr.combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "md = pbmc@meta.data # First, let's get the meta data\n",
    "i <- sapply(md, is.factor) # Identify all factor variables in your data\n",
    "md[i] <- lapply(md[i], as.character) # Convert factors to character variables\n",
    "md[is.na(md)] <- \"FALSE\" # Replace NA with \"FALSE\"\n",
    "md[i] <- lapply(md[i], as.factor) # Convert character columns back to factors\n",
    "pbmc@meta.data = md #Insert it back"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc_1 <- pbmc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Read 10X data for the second sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read 10X data:\n",
    "matrix_dir = \"~/pathway_to_data/outs/filtered_feature_bc_matrix/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "list.files(matrix_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data <- Read10X(data.dir = matrix_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "str(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create Seurat object\n",
    "pbmc <- CreateSeuratObject (counts = data, min.cells = 3, min.features = 200, project = \"pbmc_2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The number of features and UMIs (nFeature_RNA and nCount_RNA) are automatically calculated for every object by Seurat.\n",
    "# For non-UMI data, nCount_RNA represents the sum of the non-normalized values within a cell\n",
    "# We calculate the percentage of mitochondrial features here and store it in object metadata as `percent.mito`.\n",
    "# We use raw count data since this represents non-transformed and non-log-normalized counts\n",
    "# The % of UMI mapping to MT-features is a common scRNA-seq QC metric.\n",
    "mito.features <- grep(pattern = \"^MT-\", x = rownames(x = pbmc), value = TRUE)\n",
    "percent.mito <- Matrix::colSums(x = GetAssayData(object = pbmc, slot = 'counts')[mito.features, ]) / Matrix::colSums(x = GetAssayData(object = pbmc, slot = 'counts'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# The [[ operator can add columns to object metadata, and is a great place to stash QC stats\n",
    "pbmc[['percent.mito']] <- percent.mito\n",
    "VlnPlot(object = pbmc, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mito\"), ncol = 3, pt.size = 0.000001)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FeatureScatter is typically used to visualize feature-feature relationships, but can be used for anything \n",
    "# calculated by the object, i.e. columns in object metadata, PC scores etc.\n",
    "# Since there is a rare subset of cells with an outlier level of high mitochondrial percentage\n",
    "# and also low UMI content, we filter these as well\n",
    "FeatureScatter(object = pbmc, feature1 = \"nCount_RNA\", feature2 = \"percent.mito\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FeatureScatter(object = pbmc, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We filter out cells that have unique feature counts over 5,000 or less than 200\n",
    "pbmc <- subset(x = pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & percent.mito < '0.18')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add TCR information of the first sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#First we need to create TCR file for the first sample in the list\n",
    "tcr <- read.csv(paste('~/pathway_to_data/outs/', \"filtered_contig_annotations.csv\", sep=\"\"))\n",
    "tcr <- with(tcr, tcr[order(chain, decreasing = TRUE), ]) # place TRB on top before removing duplicates\n",
    "tcr <- tcr[!duplicated(tcr$barcode), ]\n",
    "#choose the columns to keep\n",
    "tcr <- tcr[,c(\"barcode\", \"raw_clonotype_id\", \"chain\", 'v_gene')]\n",
    "names(tcr)[names(tcr) == \"raw_clonotype_id\"] <- \"clonotype_id\"\n",
    "#read clonotypes file\n",
    "clono <- read.csv(paste('~/pathway_to_data/outs/', \"clonotypes.csv\", sep=\"\"))\n",
    "tcr <- merge(tcr, clono[, c(\"clonotype_id\", \"frequency\", \"cdr3s_aa\")])\n",
    "#Rename columns\n",
    "names(tcr)[1] <- \"TCR_clonotype_id\"\n",
    "names(tcr)[3] <- 'TCR_chain'\n",
    "names(tcr)[4] <- 'TCR_v_gene'\n",
    "names(tcr)[5] <- 'TCR_frequency'\n",
    "names(tcr)[6] <- 'TCR_cdr3'\n",
    "#reorder Columns\n",
    "tcr <- tcr[, c(2, 1, 3, 4, 5, 6)]\n",
    "#correct rownames\n",
    "rownames(tcr) <- tcr[,1]\n",
    "tcr[,1] <- NULL\n",
    "#Split cdr3 column:\n",
    "tcr <- separate(data = tcr, col = TCR_cdr3, into = c(\"TCR1\", \"TCR2\", \"TCR3\", \"TCR4\"), sep = \"\\\\;\")\n",
    "tcr[is.na(tcr)] <- \"FALSE\"\n",
    "head(tcr)\n",
    "tcr.combined <- tcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#divide in TRA and TRB subset:\n",
    "for (k in 1:nrow(tcr.combined)){\n",
    "  if(startsWith(tcr.combined$TCR1[k], 'TRB:')){\n",
    "    tcr.combined$TCR1B[k] <- sub(pattern = '.*:', x = tcr.combined$TCR1[k], '')\n",
    "  } else {tcr.combined$TCR1B[k] <- 'FALSE'}\n",
    "    if(startsWith(tcr.combined$TCR1[k], 'TRA:')){\n",
    "    tcr.combined$TCR1A[k] <- sub(pattern = '.*:', x = tcr.combined$TCR1[k], '')\n",
    "  } else {tcr.combined$TCR1A[k] <- 'FALSE'}\n",
    "  \n",
    "  if(startsWith(tcr.combined$TCR2[k], 'TRB:')){\n",
    "    tcr.combined$TCR2B[k] <- sub(pattern = '.*:', x = tcr.combined$TCR2[k], '')\n",
    "  } else {tcr.combined$TCR2B[k] <- 'FALSE'}\n",
    "        if(startsWith(tcr.combined$TCR2[k], 'TRA:')){\n",
    "    tcr.combined$TCR2A[k] <- sub(pattern = '.*:', x = tcr.combined$TCR2[k], '')\n",
    "  } else {tcr.combined$TCR2A[k] <- 'FALSE'}\n",
    "  \n",
    "  if(startsWith(tcr.combined$TCR3[k], 'TRB:')){\n",
    "    tcr.combined$TCR3B[k] <- sub(pattern = '.*:', x = tcr.combined$TCR3[k], '')\n",
    "  } else {tcr.combined$TCR3B[k] <- 'FALSE'}\n",
    "     if(startsWith(tcr.combined$TCR3[k], 'TRA:')){\n",
    "    tcr.combined$TCR3A[k] <- sub(pattern = '.*:', x = tcr.combined$TCR3[k], '')\n",
    "  } else {tcr.combined$TCR3A[k] <- 'FALSE'}\n",
    "  \n",
    "  if(startsWith(tcr.combined$TCR4[k], 'TRB:')){\n",
    "    tcr.combined$TCR4B[k] <- sub(pattern = '.*:', x = tcr.combined$TCR4[k], '')\n",
    "  } else {tcr.combined$TCR4B[k] <- 'FALSE'}\n",
    "    if(startsWith(tcr.combined$TCR4[k], 'TRA:')){\n",
    "    tcr.combined$TCR4A[k] <- sub(pattern = '.*:', x = tcr.combined$TCR4[k], '')\n",
    "  } else {tcr.combined$TCR4A[k] <- 'FALSE'}\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcr.combined$TCR1 <- NULL\n",
    "tcr.combined$TCR2 <- NULL\n",
    "tcr.combined$TCR3 <- NULL\n",
    "tcr.combined$TCR4 <- NULL\n",
    "head(tcr.combined)\n",
    "tail(tcr.combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nrow(tcr.combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.csv(tcr.combined, file = './tcr_2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc <- AddMetaData(object = pbmc, metadata = tcr.combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "md = pbmc@meta.data # First, let's get the meta data\n",
    "i <- sapply(md, is.factor) # Identify all factor variables in your data\n",
    "md[i] <- lapply(md[i], as.character) # Convert factors to character variables\n",
    "md[is.na(md)] <- \"FALSE\" # Replace NA with \"FALSE\"\n",
    "md[i] <- lapply(md[i], as.factor) # Convert character columns back to factors\n",
    "pbmc@meta.data = md #Insert it back"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc_2 <- pbmc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subset only T cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(pbmc.integrated@meta.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(object = pbmc.integrated) <- \"TCR_clonotype_id\"\n",
    "tcells <- subset(x = pbmc.integrated, idents = \"FALSE\", invert = TRUE)\n",
    "tcells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Integration of T cells with 2000 features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(tcells) <- 'RNA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc.list <- SplitObject(object = tcells, split.by = \"orig.ident\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pbmc.list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "markers.remove <- grep(pattern = \"^TRAV|^TRBV|^TRGV|^TRDV\",  x = rownames(x = pbmc), value = TRUE) #remove the markers for TCR variable chains before performing integration\n",
    "\n",
    "# normalize and identify variable features for each dataset independently\n",
    "pbmc.list <- lapply(X = pbmc.list, FUN = function(x) {\n",
    "    x <- NormalizeData(x)\n",
    "    x <- FindVariableFeatures(x, selection.method = \"vst\", nfeatures = 2500)\n",
    "})\n",
    "\n",
    "for (i in 1:length(pbmc.list)){\n",
    "     VariableFeatures(pbmc.list[[i]]) <- VariableFeatures(object = pbmc.list[[i]])[!(VariableFeatures(object = pbmc.list[[i]])%in%markers.remove)]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select features that are repeatedly variable across datasets for integration\n",
    "features <- SelectIntegrationFeatures(object.list = pbmc.list, nfeatures = 2000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pbmc.anchors <- FindIntegrationAnchors(object.list = pbmc.list, anchor.features = features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "tcells.integrated <- IntegrateData(anchorset = pbmc.anchors, dims = 1:30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcells.integrated\n",
    "tcells <- tcells.integrated\n",
    "tcells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(tcells, file = 'tcells_integrated_withoutumap.rds')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find VariableFeatures with vst method, scaling, dimensional reduction and clustering:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(tcells) <- 'integrated'\n",
    "tcells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "tcells <- FindVariableFeatures(tcells, selection.method = \"vst\", nfeatures = 2000)\n",
    "length(x = VariableFeatures(object = tcells))\n",
    "markers.remove <- grep(pattern = \"^TRAV|^TRBV|^TRGV|^TRDV|^RPL|^RPS\",  x = rownames(x = tcells), value = TRUE)\n",
    "VariableFeatures(object = tcells) <- VariableFeatures(object = tcells)[!(VariableFeatures(object = tcells)%in%markers.remove)]\n",
    "length(VariableFeatures(object = tcells))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcells <- ScaleData(tcells, features = VariableFeatures(object = tcells), vars.to.regress = c(\"nCount_RNA\", \"percent.mito\"))\n",
    "tcells <- RunPCA(tcells, features = VariableFeatures(object = tcells))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(tcells) <- 'orig.ident'\n",
    "DimPlot(tcells, reduction = 'pca', label = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ElbowPlot(object = tcells, ndims = 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "DimHeatmap(object = tcells, dims = 1, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 2, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 3, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 4, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 5, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 6, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 7, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 8, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 9, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 10, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 11, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 12, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 13, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 14, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 15, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 16, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 17, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 18, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 19, cells = 500, balanced = TRUE)\n",
    "DimHeatmap(object = tcells, dims = 20, cells = 500, balanced = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cluster the cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(tcells) <- 'integrated'\n",
    "tcells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcells <- FindNeighbors(object = tcells, dims = 1:18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcells <- FindClusters(object = tcells, resolution = 0.8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Non-linear dimensional reduction (tSNE) and UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "tcells <- RunUMAP(tcells, dims = 1:18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#umap, dims 18, res. 0.7\n",
    "DimPlot(tcells, reduction = 'umap', label = TRUE)\n",
    "DimPlot(tcells, reduction = 'umap', label = TRUE, group.by = 'orig.ident')\n",
    "DimPlot(tcells, reduction = 'umap', label = TRUE, group.by = 'TCR_frequency')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cluster markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(tcells) <- 'RNA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#run the DGE for each cluster\n",
    "featurestcells <- rownames(tcells)\n",
    "markers.remove <- grep(pattern = \"^TRAV|^TRBV|^TRGV|^TRDV|^RPL|^RPS\", x = rownames(tcells), value = TRUE)\n",
    "featurestcells <- featurestcells[!(featurestcells%in%markers.remove)]\n",
    "tcells.markers1 <- FindAllMarkers(object = tcells, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, features = featurestcells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#sort markers, save\n",
    "tcells.markers1_sorted <- c()\n",
    "for (i in 1:length(levels(tcells.markers1$cluster))){\n",
    "    tcells.markers1_level <- filter(tcells.markers1, cluster == levels(tcells.markers1$cluster)[i])\n",
    "    tcells.markers1_level <- tcells.markers1_level[order(-tcells.markers1_level$avg_log2FC), ]\n",
    "    tcells.markers1_level <- tcells.markers1_level[1:30, ]\n",
    "    tcells.markers1_level <- tcells.markers1_level[!is.na(tcells.markers1_level$avg_log2FC), ]\n",
    "    tcells.markers1_sorted <- rbind(tcells.markers1_sorted, tcells.markers1_level)\n",
    "    }\n",
    "tcells.markers1_sorted_top30 <- tcells.markers1_sorted\n",
    "tcells.markers1_sorted_top30\n",
    "write.csv(tcells.markers1_sorted_top30, file = './tcells_first_markers_top30.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Create average expression object\n",
    "cluster.averages_tcells <- AverageExpression(tcells, assay = \"RNA\", return.seurat = TRUE) # , verbose = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Sort markers, plot\n",
    "tcells.markers1_sorted <- c()\n",
    "for (i in 1:length(levels(tcells.markers1$cluster))){\n",
    "    tcells.markers1_level <- filter(tcells.markers1, cluster == levels(tcells.markers1$cluster)[i])\n",
    "    tcells.markers1_level <- tcells.markers1_level[order(-tcells.markers1_level$avg_log2FC), ]\n",
    "    tcells.markers1_level <- tcells.markers1_level[1:5, ]\n",
    "    tcells.markers1_level <- tcells.markers1_level[!is.na(tcells.markers1_level$avg_log2FC), ]\n",
    "    tcells.markers1_sorted <- rbind(tcells.markers1_sorted, tcells.markers1_level)\n",
    "    }\n",
    "tcells.markers1_sorted_top5 <- tcells.markers1_sorted\n",
    "tcells.markers1_sorted_top5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#look at the number of features per cell in the clusters\n",
    "VlnPlot(tcells, features = 'nFeature_RNA', pt.size = 0.01)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exploring the markers, defining the clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "levels(tcells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Print all clusters\n",
    "cluster_to_plot <- 'here the number of cluster'\n",
    "        print(DotPlot(tcells, features = filter(tcells.markers1_sorted_top30, cluster == cluster_to_plot)$gene, dot.scale = 8) + RotatedAxis() +\n",
    "                theme(axis.line = element_line(size=1),\n",
    "                text = element_text(size = 17),\n",
    "                axis.text = element_text(size = 17),\n",
    "                axis.ticks = element_line(size=1),\n",
    "                legend.text=element_text(size=17)) + ggtitle(cluster_to_plot))\n",
    "       print(cluster_to_plot)\n",
    "        \n",
    "        for(i in 1:nrow(filter(tcells.markers1_sorted_top30, cluster == cluster_to_plot))){\n",
    "             gene_to_plot <- filter(tcells.markers1_sorted_top30, cluster == cluster_to_plot)$gene[i]\n",
    "                print(VlnPlot(tcells, features = gene_to_plot, pt.size = 0.01) +\n",
    "                theme(axis.line = element_line(size=1),\n",
    "                      text = element_text(size = 20),\n",
    "                      axis.text = element_text(size = 20),\n",
    "                      axis.ticks = element_line(size=1),\n",
    "                      legend.text=element_text(size=20)) + ggtitle(paste(cluster_to_plot, ' ', gene_to_plot, sep = '')))\n",
    "                print(gene_to_plot)\n",
    "                print(FeaturePlot(tcells, features = gene_to_plot, label.size = 8, label = TRUE, pt.size = 1) + \n",
    "                theme(axis.line = element_line(size=1),\n",
    "                      text = element_text(size = 20),\n",
    "                      axis.text = element_text(size = 20),\n",
    "                      axis.ticks = element_line(size=1),\n",
    "                      legend.text=element_text(size=20))+ ggtitle(paste(cluster_to_plot, ' ', gene_to_plot, sep = '')))\n",
    "        }       "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cell cycle scoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s.genes <- cc.genes$s.genes\n",
    "g2m.genes <- cc.genes$g2m.genes\n",
    "tcells <- CellCycleScoring(tcells, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)\n",
    "head(tcells@meta.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DimPlot(tcells, reduction = 'umap', group.by = 'Phase', label = TRUE, pt.size = 0.5, label.size = 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(tcells) <- 'seurat_clusters'\n",
    "levels(tcells)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Rename the clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcells@meta.data$cd8names <- 'FALSE'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 8] <- '1_CCR7'\n",
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 3] <- '2_MAL'\n",
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 12] <- '3_NELL2'\n",
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 0] <- '4_CMC1'\n",
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 2] <- '5_FGFBP2'\n",
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 6] <- '6_FCGR3A'\n",
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 4] <- '7_LAG3'\n",
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 5] <- '8_MKI67'\n",
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 1] <- '8_MKI67'\n",
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 7] <- '9_NME1'\n",
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 10] <- '9_NME1'\n",
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 9] <- 'MAIT_1'\n",
    "tcells@meta.data$cd8names[tcells@meta.data$seurat_clusters == 11] <- 'MAIT_2'\n",
    "unique(tcells@meta.data$cd8names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(tcells) <- 'cd8names'\n",
    "levels(tcells) <- c('1_CCR7', '2_MAL', '3_NELL2', '4_CMC1', '5_FGFBP2', '6_FCGR3A', '7_LAG3', '8_MKI67', '9_NME1', 'MAIT_1', 'MAIT_2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Adding expanded column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add expand column\n",
    "for (i in 1:nrow(tcells@meta.data)){\n",
    "    if(tcells@meta.data$TCR_frequency[i] > 2){\n",
    "        tcells@meta.data$expand[i] <- 'exp'\n",
    "    } else {tcells@meta.data$expand[i] <- 'nonexp'}\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DimPlot(tcells, reduction = 'umap', group.by = 'TCR_frequency', label = TRUE)\n",
    "DimPlot(tcells, reduction = 'umap', group.by = 'expand')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc <- tcells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subset out the cells with more then one beta and alpha chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#let's find t cell doublets:\n",
    "for(i in 1:nrow(pbmc@meta.data)){\n",
    "         pbmc@meta.data$tcr_b_sum[i] <- sum(pbmc@meta.data[i, c('TCR1B', 'TCR2B', 'TCR3B', 'TCR4B')] != 'FALSE')\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save the t doublets: \n",
    "write.csv(filter(pbmc@meta.data, tcr_b_sum > 1), file = './cd8_tcelldoublets.csv')\n",
    "#save cells where only alfa chain was found:\n",
    "write.csv(filter(pbmc@meta.data, tcr_b_sum == 0), file = './cd8_tcell_no_betachain.csv')\n",
    "tcellstoremove <- c(rownames(filter(pbmc@meta.data, tcr_b_sum > 1)))\n",
    "length(tcellstoremove)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#let's find t cell doublets:\n",
    "for(i in 1:nrow(pbmc@meta.data)){\n",
    "         pbmc@meta.data$tcr_a_sum[i] <- sum(pbmc@meta.data[i, c('TCR1A', 'TCR2A', 'TCR3A', 'TCR4A')] != 'FALSE')\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save the t doublets: \n",
    "write.csv(filter(pbmc@meta.data, tcr_a_sum > 1), file = './cd8_tcelldoublets_alpha.csv')\n",
    "#save cells where only alfa chain was found:\n",
    "write.csv(filter(pbmc@meta.data, tcr_a_sum == 0), file = './cd8_tcell_no_betachain_alpa.csv')\n",
    "tcellstoremove <- c(rownames(filter(pbmc@meta.data, tcr_a_sum > 1)), tcellstoremove)\n",
    "length(tcellstoremove)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#subset everything without doublets or cells without beta chain:\n",
    "pbmc <- subset(pbmc, cells = tcellstoremove, invert = TRUE)\n",
    "pbmc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assign the clonotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clonotypes <- read.csv2('~/clonotype_info.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc@meta.data$clonotypes <- 'FALSE'\n",
    "pbmc@meta.data$specificity <- 'FALSE'\n",
    "\n",
    "#First check TCR beta\n",
    "for(i in 1:nrow(pbmc@meta.data)){\n",
    "    if(pbmc@meta.data$TCR1B[i] %in% clonotypes$CDR3.beta){\n",
    "      pbmc@meta.data$clonotypes[i] <- filter(clonotypes, CDR3.beta == pbmc@meta.data$TCR1B[i])$TCR  \n",
    "      pbmc@meta.data$specificity[i] <- filter(clonotypes, CDR3.beta == pbmc@meta.data$TCR1B[i])$Specificity  \n",
    "    }\n",
    "    \n",
    "    if(pbmc@meta.data$TCR2B[i] %in% clonotypes$CDR3.beta){\n",
    "      pbmc@meta.data$clonotypes[i] <- filter(clonotypes, CDR3.beta == pbmc@meta.data$TCR2B[i])$TCR  \n",
    "      pbmc@meta.data$specificity[i] <- filter(clonotypes, CDR3.beta == pbmc@meta.data$TCR2B[i])$Specificity  \n",
    "    }\n",
    "    \n",
    "    if(pbmc@meta.data$TCR3B[i] %in% clonotypes$CDR3.beta){\n",
    "      pbmc@meta.data$clonotypes[i] <- filter(clonotypes, CDR3.beta == pbmc@meta.data$TCR3B[i])$TCR  \n",
    "      pbmc@meta.data$specificity[i] <- filter(clonotypes, CDR3.beta == pbmc@meta.data$TCR3B[i])$Specificity  \n",
    "    }\n",
    "    \n",
    "    if(pbmc@meta.data$TCR4B[i] %in% clonotypes$CDR3.beta){\n",
    "      pbmc@meta.data$clonotypes[i] <- filter(clonotypes, CDR3.beta == pbmc@meta.data$TCR4B[i])$TCR  \n",
    "      pbmc@meta.data$specificity[i] <- filter(clonotypes, CDR3.beta == pbmc@meta.data$TCR4B[i])$Specificity  \n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#now let's check the alpha\n",
    "pbmc@meta.data$clonotypes_2 <- 'FALSE'\n",
    "for(i in 1:nrow(pbmc@meta.data)){\n",
    "    if(pbmc@meta.data$clonotypes[i] != 'FALSE'){\n",
    "        if((pbmc@meta.data$TCR1A[i] %in% clonotypes$CDR3.alpha) & (pbmc@meta.data$clonotypes[i] == filter(clonotypes, CDR3.alpha == pbmc@meta.data$TCR1A[i])$TCR[1])){\n",
    "        pbmc@meta.data$clonotypes_2[i] <- 'TRUE'\n",
    "        } \n",
    "        if((pbmc@meta.data$TCR2A[i] %in% clonotypes$CDR3.alpha) & (pbmc@meta.data$clonotypes[i] == filter(clonotypes, CDR3.alpha == pbmc@meta.data$TCR2A[i])$TCR[1])){\n",
    "        pbmc@meta.data$clonotypes_2[i] <- 'TRUE'\n",
    "        } \n",
    "        if((pbmc@meta.data$TCR3A[i] %in% clonotypes$CDR3.alpha) & (pbmc@meta.data$clonotypes[i] == filter(clonotypes, CDR3.alpha == pbmc@meta.data$TCR3A[i])$TCR[1])){\n",
    "        pbmc@meta.data$clonotypes_2[i] <- 'TRUE'\n",
    "        } \n",
    "        if((pbmc@meta.data$TCR4A[i] %in% clonotypes$CDR3.alpha) & (pbmc@meta.data$clonotypes[i] == filter(clonotypes, CDR3.alpha == pbmc@meta.data$TCR4A[i])$TCR[1])){\n",
    "        pbmc@meta.data$clonotypes_2[i] <- 'TRUE'\n",
    "        } \n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#now correct the clonotypes column\n",
    "for(i in 1:nrow(pbmc@meta.data)){\n",
    "    if(pbmc@meta.data$clonotypes_2[i] == 'FALSE'){\n",
    "        pbmc@meta.data$clonotypes[i] <- 'FALSE'\n",
    "        pbmc@meta.data$specificity[i] <- 'FALSE'\n",
    "    }\n",
    "}\n",
    "pbmc@meta.data$clonotypes_2 <- NULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#combine the H1 and H2 with the clonotypes info\n",
    "pbmc@meta.data$s_clonotypes <- 'FALSE'\n",
    "for(i in 1:nrow(pbmc@meta.data)){\n",
    "    if(pbmc@meta.data$clonotypes[i] != 'FALSE'){\n",
    "    pbmc@meta.data$s_clonotypes[i] <- paste0('H', gsub(x = pbmc@meta.data$orig.ident[i], pattern = '.*_', replacement = ''), '_', pbmc@meta.data$clonotypes[i])\n",
    "    }\n",
    "}\n",
    "unique(pbmc@meta.data$s_clonotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# unique clonotypes per sample\n",
    "\n",
    "pbmc@meta.data$sample_clono <- paste0(pbmc@meta.data$orig.ident, '_', pbmc@meta.data$TCR_clonotype_id)\n",
    "pbmc@meta.data$TCR_frequency_corrected <- 'FALSE'\n",
    "\n",
    "for(i in 1:nrow(pbmc@meta.data)){\n",
    "    pbmc@meta.data$TCR_frequency_corrected[i] <- nrow(filter(pbmc@meta.data, sample_clono == pbmc@meta.data$sample_clono[i]))\n",
    "}\n",
    "\n",
    "df <- pbmc@meta.data\n",
    "df <- df[!duplicated(df$sample_clono), ]\n",
    "rownames(df) <- 1:nrow(df)\n",
    "df <- df[!(colnames(df) %in% c('nCount_RNA', 'nFeature_RNA', 'percent.mito', 'integrated_snn_res.0.8', 'seurat_clusters', 'integrated_snn_res.0.7', 'integrated_snn_res.0.6', \n",
    "           'S.Score', 'G2M.Score', 'Phase', 'old.ident', 'cd8names', 'tcr_b_sum', 'tcr_a_sum'))]\n",
    "nrow(df)\n",
    "write.csv(df, file = 'unique_clonotypes_20221127.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#combine orig ident and speicificity column\n",
    "pbmc@meta.data$ident_specificity <- 'FALSE'\n",
    "for(i in 1:nrow(pbmc@meta.data)){\n",
    "    if(pbmc@meta.data$orig.ident[i] == 'pbmc_1'){pbmc@meta.data$ident_specificity[i] <- paste0('H1', '_', pbmc@meta.data$specificity[i])}\n",
    "    if(pbmc@meta.data$orig.ident[i] == 'pbmc_2'){pbmc@meta.data$ident_specificity[i] <- paste0('H2', '_', pbmc@meta.data$specificity[i])}\n",
    "}\n",
    "pbmc@meta.data$ident_specificity[pbmc@meta.data$ident_specificity == 'H1_FALSE'] <- 'H1_Other'\n",
    "pbmc@meta.data$ident_specificity[pbmc@meta.data$ident_specificity == 'H2_FALSE'] <- 'H2_Other'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir.create('outs_plots')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#for the UMAP clusters\n",
    "options(repr.plot.width = 14, repr.plot.height = 11)\n",
    "plot <- DimPlot(pbmc, reduction = \"umap\", label = TRUE, label.size = 8, label.box = TRUE, label.color = 'black', repel = TRUE, pt.size = 0.8, cols = brewer.pal(length(levels(pbmc)), name = 'Paired')) + \n",
    "theme(\n",
    "      text = element_text(size = 20),\n",
    "      axis.text = element_text(size = 20),\n",
    "      legend.text=element_text(size=20))\n",
    "plot\n",
    "options(repr.plot.width = 11, repr.plot.height = 11)\n",
    "ggsave(plot, file = './outs_plots/UMAP_allclusters_withlabel.eps', width = 14, height = 11)\n",
    "ggsave(plot, file = './outs_plots/UMAP_allclusters_withlabel.pdf', width = 14, height = 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#stim vs unstim\n",
    "width_p <- 13\n",
    "height_p <- 11\n",
    "dir <- './outs_plots/'\n",
    "plot_name <- 'stim_vs_unstim_umap'\n",
    "\n",
    "pbmc@meta.data$stimulation <- 'FALSE'\n",
    "pbmc@meta.data$stimulation[pbmc@meta.data$orig.ident == 'pbmc_2'] <- 'unstimulated'\n",
    "pbmc@meta.data$stimulation[pbmc@meta.data$orig.ident == 'pbmc_1'] <- 'stimulated'\n",
    "\n",
    "\n",
    "options(repr.plot.width = width_p, repr.plot.height = height_p)\n",
    "plot <- DimPlot(pbmc, reduction = \"umap\", label = FALSE, label.size = 10, repel = TRUE, group.by = 'stimulation', cols = c(\"#D3556E\", 'lightgrey'), pt.size = 0.8) + \n",
    "theme(\n",
    "      text = element_text(size = 20),\n",
    "      axis.text = element_text(size = 20),\n",
    "      legend.text=element_text(size=20))\n",
    "\n",
    "print(plot)\n",
    "ggsave(plot, file = paste0(dir, plot_name, '.pdf'), width = width_p, height = height_p)\n",
    "ggsave(plot, file = paste0(dir, plot_name, '.eps'), width = width_p, height = height_p)\n",
    "options(repr.plot.width = 11, repr.plot.height = 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#UMAP with cell stages\n",
    "options(repr.plot.width = 13, repr.plot.height = 11)\n",
    "plot <- DimPlot(pbmc, reduction = \"umap\", label = TRUE, label.size = 10, repel = TRUE, group.by = 'Phase', pt.size = 0.8) + \n",
    "theme(\n",
    "      text = element_text(size = 20),\n",
    "      axis.text = element_text(size = 20),\n",
    "      legend.text=element_text(size=20))\n",
    "plot\n",
    "ggsave(plot, file = './outs_plots/Cell_Phase_regression.eps', width = 13, height = 11)\n",
    "ggsave(plot, file = './outs_plots/Cell_Phase_regression.pdf', width = 13, height = 11)\n",
    "options(repr.plot.width = 11, repr.plot.height = 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#plot the markers\n",
    "featurespbmc <- rownames(pbmc)\n",
    "markers.remove <- grep(pattern = \"^TRAV|^TRBV|^TRGV|^TRDV|^RPL|^RPS|^IGKV|^IGLV|^IGHV|^IGHG\", x = rownames(pbmc), value = TRUE)\n",
    "featurespbmc <- featurespbmc[!(featurespbmc%in%markers.remove)]\n",
    "pbmc.markers1 <- FindAllMarkers(object = pbmc, only.pos = TRUE, min.pct = 0.2, logfc.threshold = 0.2, features = featurespbmc)\n",
    "write.csv(pbmc.markers1, file = 'markers_unsorted.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#sort markers for plotting\n",
    "pbmc.markers1_sorted <- c()\n",
    "for (i in 1:length(levels(pbmc.markers1$cluster))){\n",
    "    pbmc.markers1_level <- filter(pbmc.markers1, cluster == levels(pbmc.markers1$cluster)[i])\n",
    "    pbmc.markers1_level <- pbmc.markers1_level[order(-pbmc.markers1_level$avg_log2FC), ]\n",
    "    pbmc.markers1_level <- pbmc.markers1_level[pbmc.markers1_level$p_val_adj < 0.05, ]\n",
    "    pbmc.markers1_level <- pbmc.markers1_level[1:5, ]\n",
    "    pbmc.markers1_level <- pbmc.markers1_level[!is.na(pbmc.markers1_level$avg_log2FC), ]\n",
    "    pbmc.markers1_sorted <- rbind(pbmc.markers1_sorted, pbmc.markers1_level)\n",
    "    }\n",
    "pbmc.markers1_sorted_top5 <- pbmc.markers1_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "object <- pbmc\n",
    "levels(object) <- rev(levels(pbmc))\n",
    "cluster.averages_pbmc <- AverageExpression(object, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#build the heatmap\n",
    "options(repr.plot.width = 8, repr.plot.height = 22)\n",
    "hm <- DoHeatmap(cluster.averages_pbmc, features = pbmc.markers1_sorted_top5$gene, group.colors = rev(brewer.pal(length(levels(pbmc)), name = 'Paired')), draw.lines = FALSE,size = 8, angle = 270, hjust = 1, raster = FALSE) + theme(text = element_text(size = 12, face = \"bold\")) + \n",
    "scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\")) + theme(text = element_text(size = 25))\n",
    "ggsave(hm, file = './outs_plots/heatmap_clusters.eps', width = 8, height = 22)\n",
    "ggsave(hm, file = './outs_plots/heatmap_clusters.pdf', width = 8, height = 22)\n",
    "hm\n",
    "options(repr.plot.width = 11, repr.plot.height = 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#build the dotplot\n",
    "options(repr.plot.width = 20, repr.plot.height = 8)\n",
    "object <- pbmc\n",
    "levels(object) <- rev(levels(object))\n",
    "plot <- DotPlot(object, features = unique(pbmc.markers1_sorted_top5$gene), dot.scale = 10, cols = c('white', '#D3556E')) + RotatedAxis() +\n",
    "        theme(\n",
    "        text = element_text(size = 17),\n",
    "        axis.text = element_text(size = 17),\n",
    "        legend.text=element_text(size=17))\n",
    "plot\n",
    "ggsave(plot, file = './outs_plots/featuremap_clusters.eps', width = 20, height = 8)\n",
    "ggsave(plot, file = './outs_plots/featuremap_clusters.pdf', width = 20, height = 8)\n",
    "options(repr.plot.width = 11, repr.plot.height = 11)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot the markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "object <- pbmc\n",
    "object_average <- AverageExpression(object, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#featureplot\n",
    "markers <- c('CCR7', 'LEF1', 'NELL2', 'MAL', 'MYC', 'SELL', 'TCF7', 'NOSIP')\n",
    "\n",
    "markers <- list(markers)\n",
    "object <- AddModuleScore(object, features = markers, name = 'markers')\n",
    "plot <- FeaturePlot(object, features = 'markers1', label.size = 6, label = TRUE, pt.size = 1, repel = TRUE, cols = c(\"lightgrey\", \"#D3556E\"), min.cutoff = 'q15') + \n",
    "                theme(axis.title = element_text(size = 20),\n",
    "                        plot.title = element_text(size = 20, vjust = -5, face = 'plain'),\n",
    "                      axis.text = element_text(size = 20),\n",
    "                      legend.text=element_text(size=20))+\n",
    "                 ggtitle(paste(markers[[1]], collapse = ', '))\n",
    "plot\n",
    "ggsave(plot, file = './outs_plots/naiive_signature.eps', width = 11, height = 11)\n",
    "ggsave(plot, file = './outs_plots/naiive_signature.pdf', width = 11, height = 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#featureplot\n",
    "markers2 <- markers\n",
    "markers <- c('CX3CR1', 'GNLY', 'GZMH', 'FGFBP2', 'FCGR3A', 'PLEK', 'ADGRG1', 'PRF1')\n",
    "\n",
    "markers <- list(markers)\n",
    "object <- AddModuleScore(object, features = markers, name = 'markers')\n",
    "\n",
    "plot <- FeaturePlot(object, features = 'markers1', label.size = 6, label = TRUE, pt.size = 1, repel = TRUE, cols = c(\"lightgrey\", \"#D3556E\"), min.cutoff = 'q15') + \n",
    "                theme(axis.title = element_text(size = 20),\n",
    "                        plot.title = element_text(size = 20, vjust = -5, face = 'plain'),\n",
    "                      axis.text = element_text(size = 20),\n",
    "                      legend.text=element_text(size=20))+\n",
    "                 ggtitle(paste(markers[[1]], collapse = ', '))\n",
    "plot\n",
    "ggsave(plot, file = './outs_plots/effector_signature.eps', width = 11, height = 11)\n",
    "ggsave(plot, file = './outs_plots/effector_signature.pdf', width = 11, height = 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#heatmap\n",
    "combined <- c(markers2[[1]], markers[[1]])\n",
    "options(repr.plot.width = 8, repr.plot.height = 11)\n",
    "plot <- DoHeatmap(object_average, features = combined, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = brewer.pal(length(levels(pbmc)), name = 'Paired')) + theme(text = element_text(size = 20, face = \"bold\"))  + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "plot\n",
    "ggsave(plot, file = './outs_plots/hm_naiive_effector.eps', width = 8, height = 11)\n",
    "ggsave(plot, file = './outs_plots/hm_naiive_effector.pdf', width = 8, height = 11)\n",
    "options(repr.plot.width = 11, repr.plot.height = 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#featureplot\n",
    "markers <- c('LAG3', 'TIGIT', 'HAVCR2')\n",
    "\n",
    "markers <- list(markers)\n",
    "object <- AddModuleScore(object, features = markers, name = 'markers')\n",
    "plot <- FeaturePlot(object, features = 'markers1', label.size = 6, label = TRUE, pt.size = 1, repel = TRUE, cols = c(\"lightgrey\", \"#D3556E\"), min.cutoff = 'q15') + \n",
    "                theme(axis.title = element_text(size = 20),\n",
    "                        plot.title = element_text(size = 20, vjust = -5, face = 'plain'),\n",
    "                      axis.text = element_text(size = 20),\n",
    "                      legend.text=element_text(size=20))+\n",
    "                 ggtitle(paste(markers[[1]], collapse = ', '))\n",
    "plot\n",
    "ggsave(plot, file = './outs_plots/negative_signature.eps', width = 11, height = 11)\n",
    "ggsave(plot, file = './outs_plots/negative_signature.pdf', width = 11, height = 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#featureplot\n",
    "markers2 <- markers\n",
    "markers <- c('MKI67', 'HIST1H4C', 'HSPD1', 'NME1', 'HSP90AB1', 'ENO1', 'EIF4A1')\n",
    "\n",
    "markers <- list(markers)\n",
    "object <- AddModuleScore(object, features = markers, name = 'markers')\n",
    "\n",
    "plot <- FeaturePlot(object, features = 'markers1', label.size = 6, label = TRUE, pt.size = 1, repel = TRUE, cols = c(\"lightgrey\", \"#D3556E\"), min.cutoff = 'q15') + \n",
    "                theme(axis.title = element_text(size = 20),\n",
    "                        plot.title = element_text(size = 20, vjust = -5, face = 'plain'),\n",
    "                      axis.text = element_text(size = 20),\n",
    "                      legend.text=element_text(size=20))+\n",
    "                 ggtitle(paste(markers[[1]], collapse = ', '))\n",
    "plot\n",
    "ggsave(plot, file = './outs_plots/proliferative_signature.eps', width = 11, height = 11)\n",
    "ggsave(plot, file = './outs_plots/proliferative_signature.pdf', width = 11, height = 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#heatmap\n",
    "combined <- c(markers2[[1]], markers[[1]])\n",
    "width <- 8\n",
    "height <- 6.5\n",
    "options(repr.plot.width = width, repr.plot.height = height)\n",
    "plot <- DoHeatmap(object_average, features = combined, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = brewer.pal(length(levels(pbmc)), name = 'Paired')) + theme(text = element_text(size = 20, face = \"bold\"))  + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "plot\n",
    "ggsave(plot, file = './outs_plots/hm_negative_proliferative.eps', width = width, height = height)\n",
    "ggsave(plot, file = './outs_plots/hm_negative_proliferative.pdf', width = width, height = height)\n",
    "\n",
    "options(repr.plot.width = 11, repr.plot.height = 11)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pathway enrichment analysis per cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(httr)\n",
    "httr::set_config(config(ssl_verifypeer = FALSE))\n",
    "library('enrichR')\n",
    "setEnrichrSite(\"Enrichr\")\n",
    "dir.create('outs_clusters')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# presort the features according to the expression within the samples of the group of interest\n",
    "#define the group of analysis, cluster of interest, clustering column, patient column\n",
    "cluster_of_interest <- c('1_CCR7', '2_MAL', '3_NELL2', '4_CMC1', '5_FGFBP2', '6_FCGR3A', '7_LAG3', '8_MKI67', '9_NME1', 'MAIT_1', 'MAIT_2')\n",
    "group2 <- NULL\n",
    "object <- pbmc\n",
    "grouping <- 'cd8names'\n",
    "grouping_levels <- c('1_CCR7', '2_MAL', '3_NELL2', '4_CMC1', '5_FGFBP2', '6_FCGR3A', '7_LAG3', '8_MKI67', '9_NME1', 'MAIT_1', 'MAIT_2')\n",
    "Idents(object) <- grouping\n",
    "levels(object) <- grouping_levels\n",
    "cutoff_expression <- 0.2 #for the DGE min.pct\n",
    "logfc_cutoff <- 0.2 #for the DGE analysis\n",
    "padj_cutoff <- 0.05 #for the genes coming into the GSEA analysis\n",
    "databases_list <- c('GO_Biological_Process_2021', 'Reactome_2016')\n",
    "features <- rownames(object)\n",
    "markers.remove <- grep(pattern = \"^TRAV|^TRBV|^TRGV|^TRDV|^RPL|^RPS|^MT-|^IGKV|^IGLV|^IGHV|^IGH|^IGKC|^TRBC|^MRPL\", x = rownames(object), value = TRUE)\n",
    "features <- features[!(features%in%markers.remove)]\n",
    "#run the DGE for all clusters\n",
    "pbmc.markers <- FindAllMarkers(object = pbmc, only.pos = TRUE, min.pct = cutoff_expression, logfc.threshold = logfc_cutoff, features = features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pbmc.markers$cluster <- as.character(pbmc.markers$cluster)\n",
    "object_av <- AverageExpression(object, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "\n",
    "#here starts the loop with clusters of interest\n",
    "for(cl in 1:length(cluster_of_interest)){\n",
    "    \n",
    "    #create the dir with output\n",
    "    dir_path <- paste0('./outs_clusters/', cluster_of_interest[cl])\n",
    "    dir.create(dir_path)\n",
    "    \n",
    "    #subset the genes of interest \n",
    "    de_genes <- filter(pbmc.markers, cluster == cluster_of_interest[cl])\n",
    "    \n",
    "    #prepare the subset of genes for PEA (filter for p value)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[de_positive$p_val_adj < padj_cutoff, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #start the enrichr PEA \n",
    "    for(db in 1:length(databases_list)){\n",
    "        enriched <- enrichr(de_positive$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the positive enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_positive_', '_pea_list.csv'))\n",
    "            \n",
    "        \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_positive <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = '#d3556e') +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl],  'positive', databases_list[db]))\n",
    "            } else {\n",
    "            plot_positive <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = '#d3556e') +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl],  'positive', databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_positive, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_positive', '.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "        }\n",
    "        \n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE) + theme(text = element_text(size = 20, face = \"bold\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_HEATMAP_genes_positive', '.pdf'), width = 8, height = 11)\n",
    "        \n",
    "        vlnplot <- VlnPlot(object, features = genes_to_plot, stack = TRUE, flip = TRUE) +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))\n",
    "        ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_StackedVIOLIN_genes_positive', '.pdf'), width = 8, height = 11)\n",
    "        }\n",
    "    #end of the einrichr loop\n",
    "}\n",
    "#end of the clusters loop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Expansion overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=12, repr.plot.height=8)\n",
    "\n",
    "df <- pbmc@meta.data\n",
    "df$clusters <- pbmc$cd8names\n",
    "clusters <- unique(df$clusters)\n",
    "dis_state <- 'exp'\n",
    "df$dis <- df$expand\n",
    "df$dis <- factor(df$dis, levels = c('nonexp', 'exp'))\n",
    "\n",
    "#determine where are the most cells from HL\n",
    "order_df <- data.frame(matrix(NA, ncol = 2, nrow = length(clusters)))\n",
    "colnames(order_df) <- c('cluster', 'dis')\n",
    "order_df$cluster <- clusters\n",
    "for(i in 1:nrow(order_df)){\n",
    "    order_df$dis[i] <- 100 * nrow(filter(df, dis == dis_state & clusters == order_df$cluster[i])) / nrow(filter(df, clusters == order_df$cluster[i]))\n",
    "}\n",
    "\n",
    "order_df <- order_df[order(order_df$dis), ]\n",
    "order_list <- order_df$cluster\n",
    "\n",
    "#order the clusters in the df \n",
    "df$clusters <- factor(df$clusters, levels = order_list)\n",
    "\n",
    "#plot the type of disease state\n",
    "plot <- ggplot(df, aes_string(x=\"clusters\", fill=\"dis\")) +\n",
    "    geom_bar(position=\"fill\") + \n",
    "    theme(\n",
    "        plot.title = element_text(hjust = 0.45),\n",
    "        text = element_text(size=25),\n",
    "        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),\n",
    "        panel.background = element_blank(), axis.line = element_line(colour = \"black\"),\n",
    "        axis.text.x = element_text(angle = 55, vjust = 1, hjust=1, colour = 'black')) +\n",
    "        scale_fill_manual('legend', values = c('lightgrey', '#D3556E'))+ ylab('Fraction')+\n",
    "    ggtitle(\"Expanded cells fraction\")\n",
    "plot\n",
    "ggsave(plot, file = './outs_plots/expanded_cells_fraction.pdf', width = 12, height = 8)\n",
    "ggsave(plot, file = './outs_plots/expanded_cells_fraction.eps', width = 12, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Numbers of expanded and non-expanded cells:\n",
    "expvsnonexp <- data.frame(matrix(NA, ncol = length(unique(pbmc@meta.data$cd8names)), nrow = 5))\n",
    "colnames(expvsnonexp) <- unique(pbmc@meta.data$cd8names)\n",
    "rownames(expvsnonexp) <- c('expanded abs', 'exp relative', 'non-expanded abs', 'non-exp relative', 'total')\n",
    "for (i in 1:length(unique(pbmc@meta.data$cd8names))){\n",
    "    exp <- nrow(filter(pbmc@meta.data, expand == 'exp' & cd8names == unique(pbmc@meta.data$cd8names)[i]))\n",
    "    exprel <- 100*nrow(filter(pbmc@meta.data, expand == 'exp' & cd8names == unique(pbmc@meta.data$cd8names)[i])) / nrow(filter(pbmc@meta.data, cd8names == unique(pbmc@meta.data$cd8names)[i]))\n",
    "    nonexp <- nrow(filter(pbmc@meta.data, expand == 'nonexp' & cd8names == unique(pbmc@meta.data$cd8names)[i]))\n",
    "    nonexprel <- 100*nrow(filter(pbmc@meta.data, expand == 'nonexp' & cd8names == unique(pbmc@meta.data$cd8names)[i])) / nrow(filter(pbmc@meta.data, cd8names == unique(pbmc@meta.data$cd8names)[i]))\n",
    "    total <- nrow(filter(pbmc@meta.data, cd8names == unique(pbmc@meta.data$cd8names)[i]))\n",
    "    expvsnonexp[1, i] <- exp\n",
    "    expvsnonexp[2, i] <- exprel\n",
    "    expvsnonexp[3, i] <- nonexp\n",
    "    expvsnonexp[4, i] <- nonexprel\n",
    "    expvsnonexp[5, i] <- total            \n",
    "}        \n",
    "#order in increasing order according to expansion\n",
    "expvsnonexp1 <- expvsnonexp[, order(expvsnonexp[2, ])]\n",
    "expvsnonexp1\n",
    "write.csv(expvsnonexp1, file = './cd8expansion.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Explore in which clusters are the clonotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(RColorBrewer)\n",
    "nb.cols <- 15\n",
    "mycolors <- colorRampPalette(brewer.pal(8, \"Paired\"))(nb.cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pbmc@meta.data$s_clonotypes_2 <- pbmc@meta.data$s_clonotypes\n",
    "pbmc@meta.data$s_clonotypes[pbmc@meta.data$s_clonotypes == 'H2_4JU'] <- 'FALSE'\n",
    "pbmc@meta.data$s_clonotypes[pbmc@meta.data$s_clonotypes == 'H1_4JU'] <- 'FALSE'\n",
    "unique(pbmc@meta.data$s_clonotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#sytl4_til1 #9e1c20\n",
    "#sytl4_til2 #f36a2a\n",
    "#sytl4_pbc1 #f89d4e\n",
    "#sytl4_pbc2 #f59b9b\n",
    "\n",
    "#kif2c_pbc1 #475fad\n",
    "#kif2c_pbc2 #85b6e1\n",
    "\n",
    "#ju1 #6d08e6\n",
    "#ju2 #9568bc\n",
    "\n",
    "\n",
    "pbmc@meta.data$s_clonotypes[pbmc@meta.data$s_clonotypes == 'FALSE'] <- 'Non-assigned'\n",
    "Idents(pbmc) <- 's_clonotypes'\n",
    "levels(pbmc) <- c('H1_KIF2C-PBC1', 'H1_KIF2C-PBC2', 'H1_SYTL4-PBC1', 'H1_SYTL4-PBC2', 'H1_SYTL4-TIL1', 'H1_SYTL4-TIL2', 'H1_JU1', 'H1_JU2', 'H2_KIF2C-PBC1', 'H2_KIF2C-PBC2', 'H2_SYTL4-PBC1', 'H2_SYTL4-PBC2', 'H2_JU1', 'Non-assigned')\n",
    "colours <- c('#5cadfc', '#002ef2', '#ffa140', '#ffe0bf', '#950202', '#ff6100', '#45f000', '#1d9d00', '#5cadfc', '#002ef2', '#ffa140', '#ffe0bf', '#45f000', 'grey')\n",
    "numbers <- data.frame(matrix(NA, ncol = 2, nrow = length(levels(pbmc))))\n",
    "colnames(numbers) <- c('clonotype', 'abs_number')\n",
    "numbers$clonotype <- levels(pbmc)\n",
    "for(i in 1:nrow(numbers)){\n",
    "    numbers$abs_number[i] <- nrow(filter(pbmc@meta.data, s_clonotypes == numbers$clonotype[i]))\n",
    "}\n",
    "\n",
    "options(repr.plot.width = 14, repr.plot.height = 11)\n",
    "plot <- DimPlot(pbmc, reduction = \"umap\", label = TRUE, label.size = 7, repel = TRUE, pt.size = 0.8, cols = colours, label.box = TRUE, label.color = 'white')+ \n",
    "theme(\n",
    "      text = element_text(size = 20),\n",
    "      axis.text = element_text(size = 20),\n",
    "      legend.text=element_text(size=20)) + scale_color_manual(values = colours,\n",
    "                     labels = paste0(levels(pbmc), ' (', numbers$abs_number, ')'))\n",
    "plot \n",
    "ggsave(plot, width = 14, height = 11, file = './outs_plots/clonotypes_all_onumap.eps')\n",
    "ggsave(plot, width = 14, height = 11, file = './outs_plots/clonotypes_all_onumap.pdf')\n",
    "options(repr.plot.width = 11, repr.plot.height = 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s_clonotypes <- c('H1_KIF2C-PBC1', 'H1_KIF2C-PBC2', 'H1_SYTL4-PBC1', 'H1_SYTL4-PBC2', 'H1_SYTL4-TIL1', 'H1_SYTL4-TIL2', 'H1_JU1', 'H1_JU2', 'H2_KIF2C-PBC1', 'H2_KIF2C-PBC2', 'H2_SYTL4-PBC1', 'H2_SYTL4-PBC2', 'H2_JU1')\n",
    "clusters <- levels(pbmc)\n",
    "phases <- unique(pbmc@meta.data$Phase)\n",
    "\n",
    "columns <- levels(pbmc)\n",
    "object <- pbmc\n",
    "\n",
    "\n",
    "#build up a dataframe\n",
    "df <- data.frame(matrix(NA, ncol = length(columns), nrow = length(s_clonotypes)))\n",
    "colnames(df) <- columns\n",
    "rownames(df) <- s_clonotypes\n",
    "for(i in 1:nrow(df)){\n",
    "    for(c in 1:ncol(df)){\n",
    "        df[i, c] <- 100*nrow(filter(object@meta.data, s_clonotypes == rownames(df)[i] & cd8names == colnames(df)[c])) / nrow(filter(object@meta.data, s_clonotypes == rownames(df)[i]))\n",
    "    }\n",
    "}\n",
    "#aggregate \n",
    "aggregate <- c()\n",
    "for(i in 1:nrow(df)){\n",
    "    for(c in 1:ncol(df)){\n",
    "        result <- data.frame(rownames(df)[i], colnames(df)[c], df[i, c])\n",
    "        colnames(result) <- c('clonotype', 'column', 'result')\n",
    "        aggregate <- rbind(aggregate, result)\n",
    "    }\n",
    "}\n",
    "\n",
    "aggregate$clonotype <- factor(aggregate$clonotype, levels = s_clonotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=14, repr.plot.height=11)\n",
    "plot <- ggplot(aggregate, aes(fill=column, y=result, x=clonotype)) + \n",
    "        geom_bar(position=\"stack\", stat=\"identity\", ) + theme(\n",
    "        plot.title = element_text(hjust = 0.45),\n",
    "        text = element_text(size=30),\n",
    "        panel.grid.major = element_blank(), panel.grid.minor = element_blank(),\n",
    "        panel.background = element_blank(), axis.line = element_line(colour = \"black\"), \n",
    "        axis.text.x = element_text(angle = 55, vjust = 1, hjust=1)) + ylab('Fraction within the cluster')+ xlab('Clonotype')+\n",
    "        #scale_fill_manual(cbPalette)\n",
    "        #scale_fill_viridis(discrete = TRUE) \n",
    "        scale_fill_brewer(palette = \"Paired\")\n",
    "    print(plot)\n",
    "    ggsave(plot, width = 14, height = 11, file = './outs_plots/fraction_within_cluster.eps')\n",
    "    ggsave(plot, width = 14, height = 11, file = './outs_plots/fraction_within_cluster.pdf')\n",
    "options(repr.plot.width=11, repr.plot.height=11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colonotypes_list <- list(c('H1_KIF2C-PBC1', 'H2_KIF2C-PBC1'), \n",
    "                         c('H1_KIF2C-PBC2', 'H2_KIF2C-PBC2'),\n",
    "                        c('H1_SYTL4-PBC1', 'H2_SYTL4-PBC1'),\n",
    "                        c('H1_SYTL4-PBC2', 'H2_SYTL4-PBC2'),\n",
    "                        c('H1_SYTL4-TIL1'),\n",
    "                        c('H1_SYTL4-TIL2'),\n",
    "                        c('H1_JU1', 'H2_JU1'),\n",
    "                        c('H1_JU2')\n",
    "                        )\n",
    "colonotypes_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#stim vs unstim UMAPs for each clonotype\n",
    "width_p <- 13\n",
    "height_p <- 11\n",
    "dir <- './outs_plots/single_clonotypes_new/'\n",
    "dir.create(dir)\n",
    "\n",
    "options(repr.plot.width = width_p, repr.plot.height = height_p)\n",
    "\n",
    "clonotypes_list <- list(c('H1_KIF2C-PBC1', 'H2_KIF2C-PBC1'), \n",
    "                         c('H1_KIF2C-PBC2', 'H2_KIF2C-PBC2'),\n",
    "                        c('H1_SYTL4-PBC1', 'H2_SYTL4-PBC1'),\n",
    "                        c('H1_SYTL4-PBC2', 'H2_SYTL4-PBC2'),\n",
    "                        c('H1_SYTL4-TIL1'),\n",
    "                        c('H1_SYTL4-TIL2'),\n",
    "                        c('H1_JU1', 'H2_JU1'),\n",
    "                        c('H1_JU2')\n",
    "                        )\n",
    "\n",
    "\n",
    "for(i in 1:length(clonotypes_list)){\n",
    "    clonotypes_to_plot <- clonotypes_list[[i]]\n",
    "        \n",
    "    plot_name <- paste0(clonotypes_to_plot[1], '_stimunstim.eps')\n",
    "    \n",
    "    pbmc@meta.data$clono_plot <- 'unassigned'\n",
    "    \n",
    "    if(length(clonotypes_to_plot) > 1){\n",
    "    pbmc@meta.data$clono_plot[pbmc@meta.data$s_clonotypes == clonotypes_to_plot[1]] <- clonotypes_to_plot[1]\n",
    "    pbmc@meta.data$clono_plot[pbmc@meta.data$s_clonotypes == clonotypes_to_plot[2]] <- clonotypes_to_plot[2]\n",
    "    colors <- c('lightgrey', \"#D3556E\", '#475fad')\n",
    "    pbmc$clono_plot <- factor(pbmc$clono_plot, levels = c('unassigned', clonotypes_to_plot))\n",
    "    \n",
    "    }\n",
    "        \n",
    "    if(length(clonotypes_to_plot) == 1){\n",
    "    pbmc@meta.data$clono_plot[pbmc@meta.data$s_clonotypes == clonotypes_to_plot[1]] <- clonotypes_to_plot[1]\n",
    "    colors <- c('lightgrey', \"#D3556E\")\n",
    "    pbmc$clono_plot <- factor(pbmc$clono_plot, levels = c('unassigned', clonotypes_to_plot))\n",
    "    }\n",
    "    \n",
    "    plot <- DimPlot(pbmc, reduction = \"umap\", label = FALSE, order = TRUE, label.size = 10, repel = TRUE, group.by = 'clono_plot', cols = colors, pt.size = 2) + \n",
    "    theme(\n",
    "          text = element_text(size = 20),\n",
    "          plot.title = element_text(size = 20),\n",
    "          axis.text = element_text(size = 20),\n",
    "          legend.text=element_text(size=20)) + ggtitle(clonotypes_to_plot[1])\n",
    "    \n",
    "    print(plot)\n",
    "    ggsave(plot, file = paste0(dir, plot_name), width = width_p, height = height_p)\n",
    "        \n",
    "    plot <- DimPlot(pbmc, reduction = \"umap\", label = FALSE, order = TRUE, label.size = 10, repel = TRUE, group.by = 'clono_plot', cols = colors, pt.size = 2) + \n",
    "    theme(\n",
    "          text = element_text(size = 20),\n",
    "          plot.title = element_text(size = 20),\n",
    "          axis.text = element_text(size = 20),\n",
    "          legend.text=element_text(size=20)) + ggtitle(clonotypes_to_plot[1]) + NoLegend()\n",
    "    ggsave(plot, file = paste0(dir, 'Nolegend', plot_name), width = width_p, height = height_p)\n",
    "    \n",
    "}\n",
    "options(repr.plot.width = 11, repr.plot.height = 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#stim vs unstim\n",
    "width_p <- 13\n",
    "height_p <- 11\n",
    "dir <- './outs_plots/'\n",
    "plot_name <- 'assigned_vs_nonassigned'\n",
    "\n",
    "pbmc@meta.data$assigned <- 'non-assigned'\n",
    "pbmc@meta.data$assigned[pbmc@meta.data$clonotypes != 'FALSE'] <- 'assigned'\n",
    "\n",
    "\n",
    "options(repr.plot.width = width_p, repr.plot.height = height_p)\n",
    "plot <- DimPlot(pbmc, reduction = \"umap\", label = FALSE, label.size = 10, repel = TRUE, group.by = 'assigned', cols = c(\"#D3556E\", 'lightgrey'), pt.size = 0.8) + \n",
    "theme(\n",
    "      text = element_text(size = 20),\n",
    "      axis.text = element_text(size = 20),\n",
    "      legend.text=element_text(size=20))\n",
    "\n",
    "print(plot)\n",
    "ggsave(plot, file = paste0(dir, plot_name, '.pdf'), width = width_p, height = height_p)\n",
    "ggsave(plot, file = paste0(dir, plot_name, '.eps'), width = width_p, height = height_p)\n",
    "options(repr.plot.width = 11, repr.plot.height = 11)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DGE between the clonotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(httr)\n",
    "httr::set_config(config(ssl_verifypeer = FALSE))\n",
    "library('enrichR')\n",
    "setEnrichrSite(\"Enrichr\")\n",
    "dir.create('outs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create grouping \n",
    "\n",
    "pbmc@meta.data$h1_general <- 'FALSE'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_KIF2C')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_JU')] <- 'JU'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_SYTL4')] <- 'SYTL4'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## JU vs SYTL4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# presort the features according to the expression within the samples of the group of interest\n",
    "#define the group of analysis, cluster of interest, clustering column, patient column\n",
    "cluster_of_interest <- c('SYTL4')\n",
    "group2 <- 'JU'\n",
    "object <- pbmc\n",
    "grouping <- 'h1_general'\n",
    "grouping_levels <- c('KIF2C', 'JU', 'SYTL4', 'FALSE')\n",
    "colours_diagnosis_groups <- c('#5cadfb', '#1d9d01', '#ff8a01', '#bebebe')\n",
    "cutoff_expression <- 0.05 #for the DGE min.pct\n",
    "logfc_cutoff <- 0.02 #for the DGE analysis\n",
    "padj_cutoff <- 0.05 #for the genes coming into the GSEA analysis\n",
    "databases_list <- c('GO_Biological_Process_2021', 'Reactome_2016')\n",
    "features <- rownames(object)\n",
    "markers.remove <- grep(pattern = \"^TRAV|^TRBV|^TRGV|^TRDV|^RPL|^RPS|^MT-|^IGKV|^IGLV|^IGHV|^IGH|^IGKC|^TRBC\", x = rownames(object), value = TRUE)\n",
    "features <- features[!(features%in%markers.remove)]\n",
    "Idents(object) <- grouping\n",
    "levels(object) <- grouping_levels\n",
    "object_av <- AverageExpression(object, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "\n",
    "#for the heatmap\n",
    "object_wf <- subset(object, idents = 'FALSE', invert = TRUE)\n",
    "levels(object_wf) <- c('KIF2C', 'JU', 'SYTL4')\n",
    "object_av_wf <- AverageExpression(object_wf, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "colours_diagnosis_groups_wf <- c('#5cadfb', '#1d9d01', '#ff8a01')\n",
    "\n",
    "colour1 <- '#ff8a01'\n",
    "colour2 <- '#1d9d01'\n",
    "\n",
    "#here starts the loop with clusters of interest\n",
    "for(cl in 1:length(cluster_of_interest)){\n",
    "    \n",
    "    #create the dir with output\n",
    "    dir_path <- paste0('outs/', cluster_of_interest[cl], '_vs_', group2)\n",
    "    dir.create(dir_path)\n",
    "    \n",
    "    de_genes <- FindMarkers(object = object, features = features, only.pos = FALSE, group.by = grouping, ident.1 = cluster_of_interest[cl], ident.2 = group2, min.pct = cutoff_expression, logfc.threshold = logfc_cutoff)\n",
    "    de_genes$genes <- rownames(de_genes)\n",
    "    ##prepare the plotting table (split by the log2FC and position the most negative in reverse)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #subset negative genes\n",
    "    de_negative <- de_genes[de_genes$avg_log2FC < 0, ]\n",
    "    de_negative <- de_negative[order(de_negative$avg_log2FC), ]\n",
    "    de_genes <- rbind(de_positive, de_negative)\n",
    "    \n",
    "    #build and save the volcano plot\n",
    "    volcano <- ggplot(de_genes, aes(x = avg_log2FC, y = -log10(p_val_adj))) +\n",
    "        geom_vline(xintercept = 0) +\n",
    "        geom_hline(yintercept = -log10(padj_cutoff), color =\"grey\", linetype =\"dashed\") +\n",
    "        geom_point(data = de_genes,\n",
    "                    color = \"grey\", alpha = 0.5) +\n",
    "        geom_point(data = subset(de_genes, avg_log2FC > 0 & p_val_adj < padj_cutoff)[1:20, ],\n",
    "                    fill = colour1, alpha = 1, shape=21, size= 2.5) +\n",
    "        geom_point(data = subset(de_genes, avg_log2FC < 0 & p_val_adj < padj_cutoff)[1:20, ],\n",
    "                    fill = colour2, alpha = 1, shape=21, size= 2.5) +\n",
    "        geom_text_repel(data=rbind(subset(de_genes, avg_log2FC > 0 & p_val_adj < padj_cutoff)[1:20, ], subset(de_genes, avg_log2FC < 0 & p_val_adj < padj_cutoff)[1:20, ]), max.overlaps = 50, aes(label = genes))+\n",
    "        theme_linedraw() +\n",
    "        theme(panel.grid = element_blank(), legend.position = \"none\", \n",
    "              plot.title = element_text(size = 15, hjust = 0.5)) + ggtitle(paste0(group2, ' (left)', ' vs ', cluster_of_interest, ' (right)')) + \n",
    "        xlab(\"log2(average fold change)\") +\n",
    "        ylab(\"-log10(p-value)\")\n",
    "    ggsave(volcano, file = paste0(dir_path, '/', cluster_of_interest[cl], '_',  '_vs_', group2, '_volcano_plot.pdf'), height = 8, width = 10)\n",
    "    \n",
    "    #prepare the subset of genes for PEA (filter for p value)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[de_positive$p_val_adj < padj_cutoff, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #subset negative genes\n",
    "    de_negative <- de_genes[de_genes$avg_log2FC < 0, ]\n",
    "    de_negative <- de_negative[de_negative$p_val_adj < padj_cutoff, ]\n",
    "    de_negative <- de_negative[order(de_negative$avg_log2FC), ]\n",
    "    \n",
    "    de_all <- rbind(de_positive, de_negative)\n",
    "    #save the de_all \n",
    "    write.csv(de_all, file = paste0(dir_path, '/', cluster_of_interest[cl], '_',  '_vs_', group2, '_de_list.csv'))\n",
    "    \n",
    "    #create heatmap for positive genes\n",
    "    if(nrow(de_positive) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_positive <- object_av@assays$RNA@scale.data[de_positive$genes, ] \n",
    "    ordered_genes_positive <- as.data.frame(ordered_genes_positive)\n",
    "    ordered_genes_positive <- ordered_genes_positive[order(-ordered_genes_positive[, cluster_of_interest[cl]]), ]\n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_positive <- rownames(ordered_genes_positive)} else {ordered_genes_positive <- rownames(de_positive)} \n",
    "    \n",
    "    #top 20 sorted\n",
    "    if(length(ordered_genes_positive) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_positive[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_positive}\n",
    "    #DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7,  angle = 270, hjust = 1,\n",
    "    heatmap <- DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #build vlnplot\n",
    "    if(length(ordered_genes_plot) > 1){\n",
    "    vlnplot <- VlnPlot(object, features = ordered_genes_plot, stack = TRUE, flip = TRUE, cols = colours_diagnosis_groups, fill.by = \"ident\") +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))+ \n",
    "               geom_point_rast(position = 'jitter', size = 0.01, alpha = 0.7, colour = 'grey')\n",
    "    layer1 <- vlnplot$layers[[1]]\n",
    "    layer2 <- vlnplot$layers[[2]]\n",
    "    \n",
    "    vlnplot$layers[[1]] <- layer2\n",
    "    vlnplot$layers[[2]] <- layer1\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_vlnplot.pdf'), width = 5.5, height = 12)\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_vlnplot.eps'), width = 5.5, height = 12)}\n",
    "    \n",
    "    #create heatmap for the positive genes\n",
    "    \n",
    "    #create heatmap for positive genes\n",
    "    if(nrow(de_positive) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_positive <- object_av_wf@assays$RNA@scale.data[de_positive$genes, ] \n",
    "    ordered_genes_positive <- as.data.frame(ordered_genes_positive)\n",
    "    ordered_genes_positive <- ordered_genes_positive[order(-ordered_genes_positive[, cluster_of_interest[cl]]), ]\n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_positive <- rownames(ordered_genes_positive)} else {ordered_genes_positive <- rownames(de_positive)} \n",
    "    \n",
    "    #top 20 sorted\n",
    "    if(length(ordered_genes_positive) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_positive[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_positive}\n",
    "    #DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7,  angle = 270, hjust = 1,\n",
    "    heatmap <- DoHeatmap(object_av_wf, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups_wf) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap_without_false.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap_without_false.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #create heatmap for the negative genes\n",
    "    \n",
    "    if(nrow(de_negative) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_negative <- object_av@assays$RNA@scale.data[de_negative$genes, ] \n",
    "    ordered_genes_negative <- as.data.frame(ordered_genes_negative)\n",
    "    ordered_genes_negative <- ordered_genes_negative[order(-ordered_genes_negative[, group2]), ]\n",
    "    \n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_negative <- rownames(ordered_genes_negative)} else {ordered_genes_negative <- rownames(de_negative)}\n",
    "    \n",
    "    #top 30 sorted\n",
    "    if(length(ordered_genes_negative) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_negative[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_negative}\n",
    "    heatmap <- DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #build vlnplot\n",
    "    if(length(ordered_genes_plot) > 1){\n",
    "    vlnplot <- VlnPlot(object, features = ordered_genes_plot, stack = TRUE, flip = TRUE, cols = colours_diagnosis_groups, fill.by = \"ident\") +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))+ \n",
    "               geom_point_rast(position = 'jitter', size = 0.01, alpha = 0.7, colour = 'grey')\n",
    "    layer1 <- vlnplot$layers[[1]]\n",
    "    layer2 <- vlnplot$layers[[2]]\n",
    "    \n",
    "    vlnplot$layers[[1]] <- layer2\n",
    "    vlnplot$layers[[2]] <- layer1\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', group2, '_vlnplot.pdf'), width = 5.5, height = 12)\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', group2, '_vlnplot.eps'), width = 5.5, height = 12)}\n",
    "    \n",
    "    #create heatmap for negative genes\n",
    "    \n",
    "    if(nrow(de_negative) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_negative <- object_av_wf@assays$RNA@scale.data[de_negative$genes, ] \n",
    "    ordered_genes_negative <- as.data.frame(ordered_genes_negative)\n",
    "    ordered_genes_negative <- ordered_genes_negative[order(-ordered_genes_negative[, group2]), ]\n",
    "    \n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_negative <- rownames(ordered_genes_negative)} else {ordered_genes_negative <- rownames(de_negative)}\n",
    "    \n",
    "    #top 30 sorted\n",
    "    if(length(ordered_genes_negative) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_negative[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_negative}\n",
    "    heatmap <- DoHeatmap(object_av_wf, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups_wf) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap_without_FALSE.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap_without_FALSE.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    \n",
    "    #start the enrichr PEA \n",
    "    for(db in 1:length(databases_list)){\n",
    "        enriched <- enrichr(de_positive$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the positive enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_pea_list.csv'))\n",
    "            \n",
    "        \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        if(nrow(enriched) > 0){\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_positive <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour1) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl], databases_list[db]))\n",
    "            } else {\n",
    "            plot_positive <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour1) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl], databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_positive, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_PEA.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "        }\n",
    "        \n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE, group.colors = colours_diagnosis_groups) + theme(text = element_text(size = 20, face = \"bold\")) + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_',  '_HEATMAP_genes', '.pdf'), width = 5, height = 11)\n",
    "        \n",
    "        }\n",
    "        \n",
    "        \n",
    "        #perform the enrichment on negative genes\n",
    "        enriched <- enrichr(de_negative$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the negative enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_pea_list.csv'))\n",
    "            \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        if(nrow(enriched) > 0){\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_negative <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour2) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(group2, databases_list[db]))\n",
    "            } else {\n",
    "            plot_negative <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour2) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(group2, databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_negative, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_PEA','.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            genes_to_plot\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            genes_to_plot\n",
    "        }\n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE, group.colors = colours_diagnosis_groups) + theme(text = element_text(size = 20, face = \"bold\")) + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_HEATMAP_genes', '.pdf'), width = 5, height = 11)\n",
    "        }\n",
    "        }\n",
    "    #end of the einrichr loop\n",
    "}\n",
    "#end of the clusters loop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KIF2C_all vs SYTL4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create grouping \n",
    "\n",
    "pbmc@meta.data$h1_general_all <- 'FALSE'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_KIF2C')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_JU')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_SYTL4')] <- 'SYTL4'\n",
    "unique(pbmc@meta.data$h1_general_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique(filter(pbmc@meta.data, h1_general_all == 'KIF2C')$s_clonotypes)\n",
    "unique(filter(pbmc@meta.data, h1_general_all == 'SYTL4')$s_clonotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# presort the features according to the expression within the samples of the group of interest\n",
    "#define the group of analysis, cluster of interest, clustering column, patient column\n",
    "cluster_of_interest <- c('SYTL4')\n",
    "group2 <- 'KIF2C'\n",
    "object <- pbmc\n",
    "grouping <- 'h1_general_all'\n",
    "cutoff_expression <- 0.05 #for the DGE min.pct\n",
    "logfc_cutoff <- 0.02 #for the DGE analysis\n",
    "padj_cutoff <- 0.05 #for the genes coming into the GSEA analysis\n",
    "databases_list <- c('GO_Biological_Process_2021', 'Reactome_2016')\n",
    "features <- rownames(object)\n",
    "markers.remove <- grep(pattern = \"^TRAV|^TRBV|^TRGV|^TRDV|^RPL|^RPS|^MT-|^IGKV|^IGLV|^IGHV|^IGH|^IGKC|^TRBC\", x = rownames(object), value = TRUE)\n",
    "features <- features[!(features%in%markers.remove)]\n",
    "Idents(object) <- 'h1_general'\n",
    "levels(object) <- c('KIF2C', 'JU', 'SYTL4', 'FALSE')\n",
    "object_av <- AverageExpression(object, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "\n",
    "#for the heatmap\n",
    "object_wf <- subset(object, idents = 'FALSE', invert = TRUE)\n",
    "levels(object_wf) <- c('KIF2C', 'JU', 'SYTL4')\n",
    "object_av_wf <- AverageExpression(object_wf, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "colours_diagnosis_groups_wf <- c('#5cadfb', '#1d9d01', '#ff8a01')\n",
    "\n",
    "\n",
    "colours_diagnosis_groups <- c('#5cadfb', '#1d9d01', '#ff8a01', '#bebebe')\n",
    "colour1 <- '#ff8a01'\n",
    "colour2 <- '#12a69a'\n",
    "\n",
    "#here starts the loop with clusters of interest\n",
    "for(cl in 1:length(cluster_of_interest)){\n",
    "    \n",
    "    #create the dir with output\n",
    "    dir_path <- paste0('outs/', cluster_of_interest[cl], '_vs_', group2, '_all')\n",
    "    dir.create(dir_path)\n",
    "    \n",
    "    de_genes <- FindMarkers(object = object, features = features, only.pos = FALSE, group.by = grouping, ident.1 = cluster_of_interest[cl], ident.2 = group2, min.pct = cutoff_expression, logfc.threshold = logfc_cutoff)\n",
    "    de_genes$genes <- rownames(de_genes)\n",
    "    ##prepare the plotting table (split by the log2FC and position the most negative in reverse)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #subset negative genes\n",
    "    de_negative <- de_genes[de_genes$avg_log2FC < 0, ]\n",
    "    de_negative <- de_negative[order(de_negative$avg_log2FC), ]\n",
    "    de_genes <- rbind(de_positive, de_negative)\n",
    "    \n",
    "    #build and save the volcano plot\n",
    "    volcano <- ggplot(de_genes, aes(x = avg_log2FC, y = -log10(p_val_adj))) +\n",
    "        geom_vline(xintercept = 0) +\n",
    "        geom_hline(yintercept = -log10(padj_cutoff), color =\"grey\", linetype =\"dashed\") +\n",
    "        geom_point(data = de_genes,\n",
    "                    color = \"grey\", alpha = 0.5) +\n",
    "        geom_point(data = subset(de_genes, avg_log2FC > 0 & p_val_adj < padj_cutoff)[1:20, ],\n",
    "                    fill = colour1, alpha = 1, shape=21, size= 2.5) +\n",
    "        geom_point(data = subset(de_genes, avg_log2FC < 0 & p_val_adj < padj_cutoff)[1:20, ],\n",
    "                    fill = colour2, alpha = 1, shape=21, size= 2.5) +\n",
    "        geom_text_repel(data=rbind(subset(de_genes, avg_log2FC > 0 & p_val_adj < padj_cutoff)[1:20, ], subset(de_genes, avg_log2FC < 0 & p_val_adj < padj_cutoff)[1:20, ]), max.overlaps = 50, aes(label = genes))+\n",
    "        theme_linedraw() +\n",
    "        theme(panel.grid = element_blank(), legend.position = \"none\", \n",
    "              plot.title = element_text(size = 15, hjust = 0.5)) + ggtitle(paste0(group2, ' (left)', ' vs ', cluster_of_interest, ' (right)')) + \n",
    "        xlab(\"log2(average fold change)\") +\n",
    "        ylab(\"-log10(p-value)\")\n",
    "    ggsave(volcano, file = paste0(dir_path, '/', cluster_of_interest[cl], '_',  '_vs_', group2, '_volcano_plot.pdf'), height = 8, width = 10)\n",
    "    \n",
    "    #prepare the subset of genes for PEA (filter for p value)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[de_positive$p_val_adj < padj_cutoff, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #subset negative genes\n",
    "    de_negative <- de_genes[de_genes$avg_log2FC < 0, ]\n",
    "    de_negative <- de_negative[de_negative$p_val_adj < padj_cutoff, ]\n",
    "    de_negative <- de_negative[order(de_negative$avg_log2FC), ]\n",
    "    \n",
    "    de_all <- rbind(de_positive, de_negative)\n",
    "    #save the de_all \n",
    "    write.csv(de_all, file = paste0(dir_path, '/', cluster_of_interest[cl], '_',  '_vs_', group2, '_de_list.csv'))\n",
    "    \n",
    "    #create heatmap for positive genes\n",
    "    if(nrow(de_positive) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_positive <- object_av@assays$RNA@scale.data[de_positive$genes, ] \n",
    "    ordered_genes_positive <- as.data.frame(ordered_genes_positive)\n",
    "    ordered_genes_positive <- ordered_genes_positive[order(-ordered_genes_positive[, cluster_of_interest[cl]]), ]\n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_positive <- rownames(ordered_genes_positive)} else {ordered_genes_positive <- rownames(de_positive)} \n",
    "    \n",
    "    #top 20 sorted\n",
    "    if(length(ordered_genes_positive) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_positive[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_positive}\n",
    "    #DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7,  angle = 270, hjust = 1,\n",
    "    heatmap <- DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #build vlnplot\n",
    "    if(length(ordered_genes_plot) > 1){\n",
    "    vlnplot <- VlnPlot(object, features = ordered_genes_plot, stack = TRUE, flip = TRUE, cols = colours_diagnosis_groups, fill.by = \"ident\") +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))+ \n",
    "               geom_point_rast(position = 'jitter', size = 0.01, alpha = 0.7, colour = 'grey')\n",
    "    layer1 <- vlnplot$layers[[1]]\n",
    "    layer2 <- vlnplot$layers[[2]]\n",
    "    \n",
    "    vlnplot$layers[[1]] <- layer2\n",
    "    vlnplot$layers[[2]] <- layer1\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_vlnplot.pdf'), width = 5.5, height = 12)\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_vlnplot.eps'), width = 5.5, height = 12)}\n",
    "    \n",
    "    #create heatmap for the positive genes\n",
    "    \n",
    "    #create heatmap for positive genes\n",
    "    if(nrow(de_positive) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_positive <- object_av_wf@assays$RNA@scale.data[de_positive$genes, ] \n",
    "    ordered_genes_positive <- as.data.frame(ordered_genes_positive)\n",
    "    ordered_genes_positive <- ordered_genes_positive[order(-ordered_genes_positive[, cluster_of_interest[cl]]), ]\n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_positive <- rownames(ordered_genes_positive)} else {ordered_genes_positive <- rownames(de_positive)} \n",
    "    \n",
    "    #top 20 sorted\n",
    "    if(length(ordered_genes_positive) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_positive[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_positive}\n",
    "    #DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7,  angle = 270, hjust = 1,\n",
    "    heatmap <- DoHeatmap(object_av_wf, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups_wf) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap_without_false.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap_without_false.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #create heatmap for the negative genes\n",
    "    \n",
    "    if(nrow(de_negative) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_negative <- object_av@assays$RNA@scale.data[de_negative$genes, ] \n",
    "    ordered_genes_negative <- as.data.frame(ordered_genes_negative)\n",
    "    ordered_genes_negative <- ordered_genes_negative[order(-ordered_genes_negative[, group2]), ]\n",
    "    \n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_negative <- rownames(ordered_genes_negative)} else {ordered_genes_negative <- rownames(de_negative)}\n",
    "    \n",
    "    #top 30 sorted\n",
    "    if(length(ordered_genes_negative) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_negative[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_negative}\n",
    "    heatmap <- DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #build vlnplot\n",
    "    if(length(ordered_genes_plot) > 1){\n",
    "    vlnplot <- VlnPlot(object, features = ordered_genes_plot, stack = TRUE, flip = TRUE, cols = colours_diagnosis_groups, fill.by = \"ident\") +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))+ \n",
    "               geom_point_rast(position = 'jitter', size = 0.01, alpha = 0.7, colour = 'grey')\n",
    "    layer1 <- vlnplot$layers[[1]]\n",
    "    layer2 <- vlnplot$layers[[2]]\n",
    "    \n",
    "    vlnplot$layers[[1]] <- layer2\n",
    "    vlnplot$layers[[2]] <- layer1\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', group2, '_vlnplot.pdf'), width = 5.5, height = 12)\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', group2, '_vlnplot.eps'), width = 5.5, height = 12)}\n",
    "    \n",
    "    #create heatmap for negative genes\n",
    "    \n",
    "    if(nrow(de_negative) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_negative <- object_av_wf@assays$RNA@scale.data[de_negative$genes, ] \n",
    "    ordered_genes_negative <- as.data.frame(ordered_genes_negative)\n",
    "    ordered_genes_negative <- ordered_genes_negative[order(-ordered_genes_negative[, group2]), ]\n",
    "    \n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_negative <- rownames(ordered_genes_negative)} else {ordered_genes_negative <- rownames(de_negative)}\n",
    "    \n",
    "    #top 30 sorted\n",
    "    if(length(ordered_genes_negative) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_negative[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_negative}\n",
    "    heatmap <- DoHeatmap(object_av_wf, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups_wf) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap_without_FALSE.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap_without_FALSE.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    \n",
    "    #start the enrichr PEA \n",
    "    for(db in 1:length(databases_list)){\n",
    "        enriched <- enrichr(de_positive$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the positive enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_pea_list.csv'))\n",
    "            \n",
    "        \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        if(nrow(enriched) > 0){\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_positive <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour1) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl], databases_list[db]))\n",
    "            } else {\n",
    "            plot_positive <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour1) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl], databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_positive, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_PEA.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "        }\n",
    "        \n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE, group.colors = colours_diagnosis_groups) + theme(text = element_text(size = 20, face = \"bold\")) + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_',  '_HEATMAP_genes', '.pdf'), width = 5, height = 11)\n",
    "        \n",
    "        }\n",
    "        \n",
    "        \n",
    "        #perform the enrichment on negative genes\n",
    "        enriched <- enrichr(de_negative$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the negative enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_pea_list.csv'))\n",
    "            \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        if(nrow(enriched) > 0){\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_negative <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour2) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(group2, databases_list[db]))\n",
    "            } else {\n",
    "            plot_negative <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour2) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(group2, databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_negative, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_PEA','.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            genes_to_plot\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            genes_to_plot\n",
    "        }\n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE, group.colors = colours_diagnosis_groups) + theme(text = element_text(size = 20, face = \"bold\")) + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_HEATMAP_genes', '.pdf'), width = 5, height = 11)\n",
    "        }\n",
    "        }\n",
    "    #end of the einrichr loop\n",
    "}\n",
    "#end of the clusters loop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KIF2C_all vs JU4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create grouping dge\n",
    "\n",
    "pbmc@meta.data$h1_general_all <- 'FALSE'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_KIF2C')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_JU')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_4JU')] <- '4JU'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_SYTL4')] <- 'SYTL4'\n",
    "unique(pbmc@meta.data$h1_general_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for heatmaps\n",
    "pbmc@meta.data$h1_general <- 'FALSE'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_KIF2C')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_JU')] <- 'JU'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_4JU')] <- '4JU'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_SYTL4')] <- 'SYTL4'\n",
    "unique(pbmc@meta.data$h1_general)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique(filter(pbmc@meta.data, h1_general_all == 'KIF2C')$s_clonotypes)\n",
    "unique(filter(pbmc@meta.data, h1_general_all == '4JU')$s_clonotypes)\n",
    "unique(filter(pbmc@meta.data, h1_general_all == 'SYTL4')$s_clonotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# presort the features according to the expression within the samples of the group of interest\n",
    "#define the group of analysis, cluster of interest, clustering column, patient column\n",
    "cluster_of_interest <- c('4JU')\n",
    "group2 <- 'KIF2C'\n",
    "object <- pbmc\n",
    "grouping <- 'h1_general_all'\n",
    "cutoff_expression <- 0.05 #for the DGE min.pct\n",
    "logfc_cutoff <- 0.02 #for the DGE analysis\n",
    "padj_cutoff <- 0.05 #for the genes coming into the GSEA analysis\n",
    "databases_list <- c('GO_Biological_Process_2021', 'Reactome_2016')\n",
    "features <- rownames(object)\n",
    "markers.remove <- grep(pattern = \"^TRAV|^TRBV|^TRGV|^TRDV|^RPL|^RPS|^MT-|^IGKV|^IGLV|^IGHV|^IGH|^IGKC|^TRBC\", x = rownames(object), value = TRUE)\n",
    "features <- features[!(features%in%markers.remove)]\n",
    "\n",
    "Idents(object) <- 'h1_general'\n",
    "levels(object) <- c('KIF2C', 'JU', 'SYTL4', '4JU', 'FALSE')\n",
    "object_av <- AverageExpression(object, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "\n",
    "#for the heatmap\n",
    "object_wf <- subset(object, idents = 'FALSE', invert = TRUE)\n",
    "levels(object_wf) <- c('KIF2C', 'JU', 'SYTL4', '4JU')\n",
    "object_av_wf <- AverageExpression(object_wf, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "colours_diagnosis_groups_wf <- c('#5cadfb', '#1d9d01', '#ff8a01', '#8800ff')\n",
    "\n",
    "\n",
    "colours_diagnosis_groups <- c('#5cadfb', '#1d9d01', '#ff8a01', '#8800ff', '#bebebe')\n",
    "colour1 <- '#8800ff'\n",
    "colour2 <- '#12a69a'\n",
    "\n",
    "#here starts the loop with clusters of interest\n",
    "for(cl in 1:length(cluster_of_interest)){\n",
    "    \n",
    "    #create the dir with output\n",
    "    dir_path <- paste0('outs/', cluster_of_interest[cl], '_vs_', group2, '_all')\n",
    "    dir.create(dir_path)\n",
    "    \n",
    "    de_genes <- FindMarkers(object = object, features = features, only.pos = FALSE, group.by = grouping, ident.1 = cluster_of_interest[cl], ident.2 = group2, min.pct = cutoff_expression, logfc.threshold = logfc_cutoff)\n",
    "    de_genes$genes <- rownames(de_genes)\n",
    "    ##prepare the plotting table (split by the log2FC and position the most negative in reverse)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #subset negative genes\n",
    "    de_negative <- de_genes[de_genes$avg_log2FC < 0, ]\n",
    "    de_negative <- de_negative[order(de_negative$avg_log2FC), ]\n",
    "    de_genes <- rbind(de_positive, de_negative)\n",
    "    \n",
    "    #build and save the volcano plot\n",
    "    volcano <- ggplot(de_genes, aes(x = avg_log2FC, y = -log10(p_val_adj))) +\n",
    "        geom_vline(xintercept = 0) +\n",
    "        geom_hline(yintercept = -log10(padj_cutoff), color =\"grey\", linetype =\"dashed\") +\n",
    "        geom_point(data = de_genes,\n",
    "                    color = \"grey\", alpha = 0.5) +\n",
    "        geom_point(data = subset(de_genes, avg_log2FC > 0 & p_val_adj < padj_cutoff)[1:20, ],\n",
    "                    fill = colour1, alpha = 1, shape=21, size= 2.5) +\n",
    "        geom_point(data = subset(de_genes, avg_log2FC < 0 & p_val_adj < padj_cutoff)[1:20, ],\n",
    "                    fill = colour2, alpha = 1, shape=21, size= 2.5) +\n",
    "        geom_text_repel(data=rbind(subset(de_genes, avg_log2FC > 0 & p_val_adj < padj_cutoff)[1:20, ], subset(de_genes, avg_log2FC < 0 & p_val_adj < padj_cutoff)[1:20, ]), max.overlaps = 50, aes(label = genes))+\n",
    "        theme_linedraw() +\n",
    "        theme(panel.grid = element_blank(), legend.position = \"none\", \n",
    "              plot.title = element_text(size = 15, hjust = 0.5)) + ggtitle(paste0(group2, ' (left)', ' vs ', cluster_of_interest, ' (right)')) + \n",
    "        xlab(\"log2(average fold change)\") +\n",
    "        ylab(\"-log10(p-value)\")\n",
    "    ggsave(volcano, file = paste0(dir_path, '/', cluster_of_interest[cl], '_',  '_vs_', group2, '_volcano_plot.pdf'), height = 8, width = 10)\n",
    "    \n",
    "    #prepare the subset of genes for PEA (filter for p value)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[de_positive$p_val_adj < padj_cutoff, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #subset negative genes\n",
    "    de_negative <- de_genes[de_genes$avg_log2FC < 0, ]\n",
    "    de_negative <- de_negative[de_negative$p_val_adj < padj_cutoff, ]\n",
    "    de_negative <- de_negative[order(de_negative$avg_log2FC), ]\n",
    "    \n",
    "    de_all <- rbind(de_positive, de_negative)\n",
    "    #save the de_all \n",
    "    write.csv(de_all, file = paste0(dir_path, '/', cluster_of_interest[cl], '_',  '_vs_', group2, '_de_list.csv'))\n",
    "    \n",
    "    #create heatmap for positive genes\n",
    "    if(nrow(de_positive) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_positive <- object_av@assays$RNA@scale.data[de_positive$genes, ] \n",
    "    ordered_genes_positive <- as.data.frame(ordered_genes_positive)\n",
    "    ordered_genes_positive <- ordered_genes_positive[order(-ordered_genes_positive[, cluster_of_interest[cl]]), ]\n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_positive <- rownames(ordered_genes_positive)} else {ordered_genes_positive <- rownames(de_positive)} \n",
    "    \n",
    "    #top 20 sorted\n",
    "    if(length(ordered_genes_positive) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_positive[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_positive}\n",
    "    #DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7,  angle = 270, hjust = 1,\n",
    "    heatmap <- DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #build vlnplot\n",
    "    if(length(ordered_genes_plot) > 1){\n",
    "    vlnplot <- VlnPlot(object, features = ordered_genes_plot, stack = TRUE, flip = TRUE, cols = colours_diagnosis_groups, fill.by = \"ident\") +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))+ \n",
    "               geom_point_rast(position = 'jitter', size = 0.01, alpha = 0.7, colour = 'grey')\n",
    "    layer1 <- vlnplot$layers[[1]]\n",
    "    layer2 <- vlnplot$layers[[2]]\n",
    "    \n",
    "    vlnplot$layers[[1]] <- layer2\n",
    "    vlnplot$layers[[2]] <- layer1\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_vlnplot.pdf'), width = 5.5, height = 12)\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_vlnplot.eps'), width = 5.5, height = 12)}\n",
    "    \n",
    "    #create heatmap for the positive genes\n",
    "    \n",
    "    #create heatmap for positive genes\n",
    "    if(nrow(de_positive) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_positive <- object_av_wf@assays$RNA@scale.data[de_positive$genes, ] \n",
    "    ordered_genes_positive <- as.data.frame(ordered_genes_positive)\n",
    "    ordered_genes_positive <- ordered_genes_positive[order(-ordered_genes_positive[, cluster_of_interest[cl]]), ]\n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_positive <- rownames(ordered_genes_positive)} else {ordered_genes_positive <- rownames(de_positive)} \n",
    "    \n",
    "    #top 20 sorted\n",
    "    if(length(ordered_genes_positive) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_positive[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_positive}\n",
    "    #DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7,  angle = 270, hjust = 1,\n",
    "    heatmap <- DoHeatmap(object_av_wf, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups_wf) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap_without_false.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap_without_false.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #create heatmap for the negative genes\n",
    "    \n",
    "    if(nrow(de_negative) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_negative <- object_av@assays$RNA@scale.data[de_negative$genes, ] \n",
    "    ordered_genes_negative <- as.data.frame(ordered_genes_negative)\n",
    "    ordered_genes_negative <- ordered_genes_negative[order(-ordered_genes_negative[, group2]), ]\n",
    "    \n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_negative <- rownames(ordered_genes_negative)} else {ordered_genes_negative <- rownames(de_negative)}\n",
    "    \n",
    "    #top 30 sorted\n",
    "    if(length(ordered_genes_negative) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_negative[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_negative}\n",
    "    heatmap <- DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #build vlnplot\n",
    "    if(length(ordered_genes_plot) > 1){\n",
    "    vlnplot <- VlnPlot(object, features = ordered_genes_plot, stack = TRUE, flip = TRUE, cols = colours_diagnosis_groups, fill.by = \"ident\") +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))+ \n",
    "               geom_point_rast(position = 'jitter', size = 0.01, alpha = 0.7, colour = 'grey')\n",
    "    layer1 <- vlnplot$layers[[1]]\n",
    "    layer2 <- vlnplot$layers[[2]]\n",
    "    \n",
    "    vlnplot$layers[[1]] <- layer2\n",
    "    vlnplot$layers[[2]] <- layer1\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', group2, '_vlnplot.pdf'), width = 5.5, height = 12)\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', group2, '_vlnplot.eps'), width = 5.5, height = 12)}\n",
    "    \n",
    "    #create heatmap for negative genes\n",
    "    \n",
    "    if(nrow(de_negative) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_negative <- object_av_wf@assays$RNA@scale.data[de_negative$genes, ] \n",
    "    ordered_genes_negative <- as.data.frame(ordered_genes_negative)\n",
    "    ordered_genes_negative <- ordered_genes_negative[order(-ordered_genes_negative[, group2]), ]\n",
    "    \n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_negative <- rownames(ordered_genes_negative)} else {ordered_genes_negative <- rownames(de_negative)}\n",
    "    \n",
    "    #top 30 sorted\n",
    "    if(length(ordered_genes_negative) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_negative[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_negative}\n",
    "    heatmap <- DoHeatmap(object_av_wf, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups_wf) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap_without_FALSE.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap_without_FALSE.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    \n",
    "    #start the enrichr PEA \n",
    "    for(db in 1:length(databases_list)){\n",
    "        enriched <- enrichr(de_positive$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the positive enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_pea_list.csv'))\n",
    "            \n",
    "        \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        if(nrow(enriched) > 0){\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_positive <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour1) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl], databases_list[db]))\n",
    "            } else {\n",
    "            plot_positive <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour1) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl], databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_positive, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_PEA.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "        }\n",
    "        \n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE, group.colors = colours_diagnosis_groups) + theme(text = element_text(size = 20, face = \"bold\")) + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_',  '_HEATMAP_genes', '.pdf'), width = 5, height = 11)\n",
    "        \n",
    "        }\n",
    "        \n",
    "        \n",
    "        #perform the enrichment on negative genes\n",
    "        enriched <- enrichr(de_negative$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the negative enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_pea_list.csv'))\n",
    "            \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        if(nrow(enriched) > 0){\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_negative <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour2) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(group2, databases_list[db]))\n",
    "            } else {\n",
    "            plot_negative <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour2) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(group2, databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_negative, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_PEA','.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            genes_to_plot\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            genes_to_plot\n",
    "        }\n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE, group.colors = colours_diagnosis_groups) + theme(text = element_text(size = 20, face = \"bold\")) + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_HEATMAP_genes', '.pdf'), width = 5, height = 11)\n",
    "        }\n",
    "        }\n",
    "    #end of the einrichr loop\n",
    "}\n",
    "#end of the clusters loop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SYTL4 vs JU4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create grouping dge\n",
    "\n",
    "pbmc@meta.data$h1_general_all <- 'FALSE'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_KIF2C')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_JU')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_4JU')] <- '4JU'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_SYTL4')] <- 'SYTL4'\n",
    "unique(pbmc@meta.data$h1_general_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for heatmaps\n",
    "pbmc@meta.data$h1_general <- 'FALSE'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_KIF2C')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_JU')] <- 'JU'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_4JU')] <- '4JU'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_SYTL4')] <- 'SYTL4'\n",
    "unique(pbmc@meta.data$h1_general)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# presort the features according to the expression within the samples of the group of interest\n",
    "#define the group of analysis, cluster of interest, clustering column, patient column\n",
    "cluster_of_interest <- c('4JU')\n",
    "group2 <- 'SYTL4'\n",
    "object <- pbmc\n",
    "grouping <- 'h1_general_all'\n",
    "cutoff_expression <- 0.05 #for the DGE min.pct\n",
    "logfc_cutoff <- 0.02 #for the DGE analysis\n",
    "padj_cutoff <- 0.05 #for the genes coming into the GSEA analysis\n",
    "databases_list <- c('GO_Biological_Process_2021', 'Reactome_2016')\n",
    "features <- rownames(object)\n",
    "markers.remove <- grep(pattern = \"^TRAV|^TRBV|^TRGV|^TRDV|^RPL|^RPS|^MT-|^IGKV|^IGLV|^IGHV|^IGH|^IGKC|^TRBC\", x = rownames(object), value = TRUE)\n",
    "features <- features[!(features%in%markers.remove)]\n",
    "\n",
    "Idents(object) <- 'h1_general'\n",
    "levels(object) <- c('KIF2C', 'JU', 'SYTL4', '4JU', 'FALSE')\n",
    "object_av <- AverageExpression(object, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "\n",
    "#for the heatmap\n",
    "object_wf <- subset(object, idents = 'FALSE', invert = TRUE)\n",
    "levels(object_wf) <- c('KIF2C', 'JU', 'SYTL4', '4JU')\n",
    "object_av_wf <- AverageExpression(object_wf, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "colours_diagnosis_groups_wf <- c('#5cadfb', '#1d9d01', '#ff8a01', '#8800ff')\n",
    "\n",
    "\n",
    "colours_diagnosis_groups <- c('#5cadfb', '#1d9d01', '#ff8a01', '#8800ff', '#bebebe')\n",
    "colour1 <- '#8800ff'\n",
    "colour2 <- '#ff8a01'\n",
    "\n",
    "#here starts the loop with clusters of interest\n",
    "for(cl in 1:length(cluster_of_interest)){\n",
    "    \n",
    "    #create the dir with output\n",
    "    dir_path <- paste0('outs/', cluster_of_interest[cl], '_vs_', group2, '_all')\n",
    "    dir.create(dir_path)\n",
    "    \n",
    "    de_genes <- FindMarkers(object = object, features = features, only.pos = FALSE, group.by = grouping, ident.1 = cluster_of_interest[cl], ident.2 = group2, min.pct = cutoff_expression, logfc.threshold = logfc_cutoff)\n",
    "    de_genes$genes <- rownames(de_genes)\n",
    "    ##prepare the plotting table (split by the log2FC and position the most negative in reverse)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #subset negative genes\n",
    "    de_negative <- de_genes[de_genes$avg_log2FC < 0, ]\n",
    "    de_negative <- de_negative[order(de_negative$avg_log2FC), ]\n",
    "    de_genes <- rbind(de_positive, de_negative)\n",
    "    \n",
    "    #build and save the volcano plot\n",
    "    volcano <- ggplot(de_genes, aes(x = avg_log2FC, y = -log10(p_val_adj))) +\n",
    "        geom_vline(xintercept = 0) +\n",
    "        geom_hline(yintercept = -log10(padj_cutoff), color =\"grey\", linetype =\"dashed\") +\n",
    "        geom_point(data = de_genes,\n",
    "                    color = \"grey\", alpha = 0.5) +\n",
    "        geom_point(data = subset(de_genes, avg_log2FC > 0 & p_val_adj < padj_cutoff)[1:20, ],\n",
    "                    fill = colour1, alpha = 1, shape=21, size= 2.5) +\n",
    "        geom_point(data = subset(de_genes, avg_log2FC < 0 & p_val_adj < padj_cutoff)[1:20, ],\n",
    "                    fill = colour2, alpha = 1, shape=21, size= 2.5) +\n",
    "        geom_text_repel(data=rbind(subset(de_genes, avg_log2FC > 0 & p_val_adj < padj_cutoff)[1:20, ], subset(de_genes, avg_log2FC < 0 & p_val_adj < padj_cutoff)[1:20, ]), max.overlaps = 50, aes(label = genes))+\n",
    "        theme_linedraw() +\n",
    "        theme(panel.grid = element_blank(), legend.position = \"none\", \n",
    "              plot.title = element_text(size = 15, hjust = 0.5)) + ggtitle(paste0(group2, ' (left)', ' vs ', cluster_of_interest, ' (right)')) + \n",
    "        xlab(\"log2(average fold change)\") +\n",
    "        ylab(\"-log10(p-value)\")\n",
    "    ggsave(volcano, file = paste0(dir_path, '/', cluster_of_interest[cl], '_',  '_vs_', group2, '_volcano_plot.pdf'), height = 8, width = 10)\n",
    "    \n",
    "    #prepare the subset of genes for PEA (filter for p value)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[de_positive$p_val_adj < padj_cutoff, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #subset negative genes\n",
    "    de_negative <- de_genes[de_genes$avg_log2FC < 0, ]\n",
    "    de_negative <- de_negative[de_negative$p_val_adj < padj_cutoff, ]\n",
    "    de_negative <- de_negative[order(de_negative$avg_log2FC), ]\n",
    "    \n",
    "    de_all <- rbind(de_positive, de_negative)\n",
    "    #save the de_all \n",
    "    write.csv(de_all, file = paste0(dir_path, '/', cluster_of_interest[cl], '_',  '_vs_', group2, '_de_list.csv'))\n",
    "    \n",
    "    #create heatmap for positive genes\n",
    "    if(nrow(de_positive) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_positive <- object_av@assays$RNA@scale.data[de_positive$genes, ] \n",
    "    ordered_genes_positive <- as.data.frame(ordered_genes_positive)\n",
    "    ordered_genes_positive <- ordered_genes_positive[order(-ordered_genes_positive[, cluster_of_interest[cl]]), ]\n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_positive <- rownames(ordered_genes_positive)} else {ordered_genes_positive <- rownames(de_positive)} \n",
    "    \n",
    "    #top 20 sorted\n",
    "    if(length(ordered_genes_positive) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_positive[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_positive}\n",
    "    #DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7,  angle = 270, hjust = 1,\n",
    "    heatmap <- DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #build vlnplot\n",
    "    if(length(ordered_genes_plot) > 1){\n",
    "    vlnplot <- VlnPlot(object, features = ordered_genes_plot, stack = TRUE, flip = TRUE, cols = colours_diagnosis_groups, fill.by = \"ident\") +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))+ \n",
    "               geom_point_rast(position = 'jitter', size = 0.01, alpha = 0.7, colour = 'grey')\n",
    "    layer1 <- vlnplot$layers[[1]]\n",
    "    layer2 <- vlnplot$layers[[2]]\n",
    "    \n",
    "    vlnplot$layers[[1]] <- layer2\n",
    "    vlnplot$layers[[2]] <- layer1\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_vlnplot.pdf'), width = 5.5, height = 12)\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_vlnplot.eps'), width = 5.5, height = 12)}\n",
    "    \n",
    "    #create heatmap for the positive genes\n",
    "    \n",
    "    #create heatmap for positive genes\n",
    "    if(nrow(de_positive) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_positive <- object_av_wf@assays$RNA@scale.data[de_positive$genes, ] \n",
    "    ordered_genes_positive <- as.data.frame(ordered_genes_positive)\n",
    "    ordered_genes_positive <- ordered_genes_positive[order(-ordered_genes_positive[, cluster_of_interest[cl]]), ]\n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_positive <- rownames(ordered_genes_positive)} else {ordered_genes_positive <- rownames(de_positive)} \n",
    "    \n",
    "    #top 20 sorted\n",
    "    if(length(ordered_genes_positive) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_positive[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_positive}\n",
    "    #DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7,  angle = 270, hjust = 1,\n",
    "    heatmap <- DoHeatmap(object_av_wf, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups_wf) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap_without_false.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap_without_false.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #create heatmap for the negative genes\n",
    "    \n",
    "    if(nrow(de_negative) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_negative <- object_av@assays$RNA@scale.data[de_negative$genes, ] \n",
    "    ordered_genes_negative <- as.data.frame(ordered_genes_negative)\n",
    "    ordered_genes_negative <- ordered_genes_negative[order(-ordered_genes_negative[, group2]), ]\n",
    "    \n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_negative <- rownames(ordered_genes_negative)} else {ordered_genes_negative <- rownames(de_negative)}\n",
    "    \n",
    "    #top 30 sorted\n",
    "    if(length(ordered_genes_negative) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_negative[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_negative}\n",
    "    heatmap <- DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #build vlnplot\n",
    "    if(length(ordered_genes_plot) > 1){\n",
    "    vlnplot <- VlnPlot(object, features = ordered_genes_plot, stack = TRUE, flip = TRUE, cols = colours_diagnosis_groups, fill.by = \"ident\") +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))+ \n",
    "               geom_point_rast(position = 'jitter', size = 0.01, alpha = 0.7, colour = 'grey')\n",
    "    layer1 <- vlnplot$layers[[1]]\n",
    "    layer2 <- vlnplot$layers[[2]]\n",
    "    \n",
    "    vlnplot$layers[[1]] <- layer2\n",
    "    vlnplot$layers[[2]] <- layer1\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', group2, '_vlnplot.pdf'), width = 5.5, height = 12)\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', group2, '_vlnplot.eps'), width = 5.5, height = 12)}\n",
    "    \n",
    "    #create heatmap for negative genes\n",
    "    \n",
    "    if(nrow(de_negative) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_negative <- object_av_wf@assays$RNA@scale.data[de_negative$genes, ] \n",
    "    ordered_genes_negative <- as.data.frame(ordered_genes_negative)\n",
    "    ordered_genes_negative <- ordered_genes_negative[order(-ordered_genes_negative[, group2]), ]\n",
    "    \n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_negative <- rownames(ordered_genes_negative)} else {ordered_genes_negative <- rownames(de_negative)}\n",
    "    \n",
    "    #top 30 sorted\n",
    "    if(length(ordered_genes_negative) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_negative[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_negative}\n",
    "    heatmap <- DoHeatmap(object_av_wf, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups_wf) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap_without_FALSE.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap_without_FALSE.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    \n",
    "    #start the enrichr PEA \n",
    "    for(db in 1:length(databases_list)){\n",
    "        enriched <- enrichr(de_positive$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the positive enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_pea_list.csv'))\n",
    "            \n",
    "        \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        if(nrow(enriched) > 0){\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_positive <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour1) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl], databases_list[db]))\n",
    "            } else {\n",
    "            plot_positive <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour1) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl], databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_positive, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_PEA.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "        }\n",
    "        \n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE, group.colors = colours_diagnosis_groups) + theme(text = element_text(size = 20, face = \"bold\")) + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_',  '_HEATMAP_genes', '.pdf'), width = 5, height = 11)\n",
    "        \n",
    "        }\n",
    "        \n",
    "        \n",
    "        #perform the enrichment on negative genes\n",
    "        enriched <- enrichr(de_negative$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the negative enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_pea_list.csv'))\n",
    "            \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        if(nrow(enriched) > 0){\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_negative <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour2) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(group2, databases_list[db]))\n",
    "            } else {\n",
    "            plot_negative <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour2) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(group2, databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_negative, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_PEA','.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            genes_to_plot\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            genes_to_plot\n",
    "        }\n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE, group.colors = colours_diagnosis_groups) + theme(text = element_text(size = 20, face = \"bold\")) + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_HEATMAP_genes', '.pdf'), width = 5, height = 11)\n",
    "        }\n",
    "        }\n",
    "    #end of the einrichr loop\n",
    "}\n",
    "#end of the clusters loop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KIF2C_all vs JU4/JU3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create grouping dge\n",
    "\n",
    "pbmc@meta.data$h1_general_all <- 'FALSE'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_KIF2C')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_JU')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_4JU') | startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_3JU')] <- '3or4JU'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_SYTL4')] <- 'SYTL4'\n",
    "unique(pbmc@meta.data$h1_general_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for heatmaps\n",
    "pbmc@meta.data$h1_general <- 'FALSE'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_KIF2C')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_JU')] <- 'JU'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_4JU') | startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_3JU')] <- '3or4JU'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_SYTL4')] <- 'SYTL4'\n",
    "unique(pbmc@meta.data$h1_general)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# presort the features according to the expression within the samples of the group of interest\n",
    "#define the group of analysis, cluster of interest, clustering column, patient column\n",
    "cluster_of_interest <- c('3or4JU')\n",
    "group2 <- 'KIF2C'\n",
    "object <- pbmc\n",
    "grouping <- 'h1_general_all'\n",
    "cutoff_expression <- 0.05 #for the DGE min.pct\n",
    "logfc_cutoff <- 0.02 #for the DGE analysis\n",
    "padj_cutoff <- 0.05 #for the genes coming into the GSEA analysis\n",
    "databases_list <- c('GO_Biological_Process_2021', 'Reactome_2016')\n",
    "features <- rownames(object)\n",
    "markers.remove <- grep(pattern = \"^TRAV|^TRBV|^TRGV|^TRDV|^RPL|^RPS|^MT-|^IGKV|^IGLV|^IGHV|^IGH|^IGKC|^TRBC\", x = rownames(object), value = TRUE)\n",
    "features <- features[!(features%in%markers.remove)]\n",
    "\n",
    "Idents(object) <- 'h1_general'\n",
    "levels(object) <- c('KIF2C', 'JU', 'SYTL4', '3or4JU', 'FALSE')\n",
    "object_av <- AverageExpression(object, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "\n",
    "#for the heatmap\n",
    "object_wf <- subset(object, idents = 'FALSE', invert = TRUE)\n",
    "levels(object_wf) <- c('KIF2C', 'JU', 'SYTL4', '3or4JU')\n",
    "object_av_wf <- AverageExpression(object_wf, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "colours_diagnosis_groups_wf <- c('#5cadfb', '#1d9d01', '#ff8a01', '#8800ff')\n",
    "\n",
    "\n",
    "colours_diagnosis_groups <- c('#5cadfb', '#1d9d01', '#ff8a01', '#8800ff', '#bebebe')\n",
    "colour1 <- '#8800ff'\n",
    "colour2 <- '#12a69a'\n",
    "\n",
    "#here starts the loop with clusters of interest\n",
    "for(cl in 1:length(cluster_of_interest)){\n",
    "    \n",
    "    #create the dir with output\n",
    "    dir_path <- paste0('outs/', cluster_of_interest[cl], '_vs_', group2, '_all')\n",
    "    dir.create(dir_path)\n",
    "    \n",
    "    de_genes <- FindMarkers(object = object, features = features, only.pos = FALSE, group.by = grouping, ident.1 = cluster_of_interest[cl], ident.2 = group2, min.pct = cutoff_expression, logfc.threshold = logfc_cutoff)\n",
    "    de_genes$genes <- rownames(de_genes)\n",
    "    ##prepare the plotting table (split by the log2FC and position the most negative in reverse)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #subset negative genes\n",
    "    de_negative <- de_genes[de_genes$avg_log2FC < 0, ]\n",
    "    de_negative <- de_negative[order(de_negative$avg_log2FC), ]\n",
    "    de_genes <- rbind(de_positive, de_negative)\n",
    "    \n",
    "    #build and save the volcano plot\n",
    "    volcano <- ggplot(de_genes, aes(x = avg_log2FC, y = -log10(p_val_adj))) +\n",
    "        geom_vline(xintercept = 0) +\n",
    "        geom_hline(yintercept = -log10(padj_cutoff), color =\"grey\", linetype =\"dashed\") +\n",
    "        geom_point(data = de_genes,\n",
    "                    color = \"grey\", alpha = 0.5) +\n",
    "        geom_point(data = subset(de_genes, avg_log2FC > 0 & p_val_adj < padj_cutoff)[1:20, ],\n",
    "                    fill = colour1, alpha = 1, shape=21, size= 2.5) +\n",
    "        geom_point(data = subset(de_genes, avg_log2FC < 0 & p_val_adj < padj_cutoff)[1:20, ],\n",
    "                    fill = colour2, alpha = 1, shape=21, size= 2.5) +\n",
    "        geom_text_repel(data=rbind(subset(de_genes, avg_log2FC > 0 & p_val_adj < padj_cutoff)[1:20, ], subset(de_genes, avg_log2FC < 0 & p_val_adj < padj_cutoff)[1:20, ]), max.overlaps = 50, aes(label = genes))+\n",
    "        theme_linedraw() +\n",
    "        theme(panel.grid = element_blank(), legend.position = \"none\", \n",
    "              plot.title = element_text(size = 15, hjust = 0.5)) + ggtitle(paste0(group2, ' (left)', ' vs ', cluster_of_interest, ' (right)')) + \n",
    "        xlab(\"log2(average fold change)\") +\n",
    "        ylab(\"-log10(p-value)\")\n",
    "    ggsave(volcano, file = paste0(dir_path, '/', cluster_of_interest[cl], '_',  '_vs_', group2, '_volcano_plot.pdf'), height = 8, width = 10)\n",
    "    \n",
    "    #prepare the subset of genes for PEA (filter for p value)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[de_positive$p_val_adj < padj_cutoff, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #subset negative genes\n",
    "    de_negative <- de_genes[de_genes$avg_log2FC < 0, ]\n",
    "    de_negative <- de_negative[de_negative$p_val_adj < padj_cutoff, ]\n",
    "    de_negative <- de_negative[order(de_negative$avg_log2FC), ]\n",
    "    \n",
    "    de_all <- rbind(de_positive, de_negative)\n",
    "    #save the de_all \n",
    "    write.csv(de_all, file = paste0(dir_path, '/', cluster_of_interest[cl], '_',  '_vs_', group2, '_de_list.csv'))\n",
    "    \n",
    "    #create heatmap for positive genes\n",
    "    if(nrow(de_positive) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_positive <- object_av@assays$RNA@scale.data[de_positive$genes, ] \n",
    "    ordered_genes_positive <- as.data.frame(ordered_genes_positive)\n",
    "    ordered_genes_positive <- ordered_genes_positive[order(-ordered_genes_positive[, cluster_of_interest[cl]]), ]\n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_positive <- rownames(ordered_genes_positive)} else {ordered_genes_positive <- rownames(de_positive)} \n",
    "    \n",
    "    #top 20 sorted\n",
    "    if(length(ordered_genes_positive) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_positive[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_positive}\n",
    "    #DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7,  angle = 270, hjust = 1,\n",
    "    heatmap <- DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #build vlnplot\n",
    "    if(length(ordered_genes_plot) > 1){\n",
    "    vlnplot <- VlnPlot(object, features = ordered_genes_plot, stack = TRUE, flip = TRUE, cols = colours_diagnosis_groups, fill.by = \"ident\") +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))+ \n",
    "               geom_point_rast(position = 'jitter', size = 0.01, alpha = 0.7, colour = 'grey')\n",
    "    layer1 <- vlnplot$layers[[1]]\n",
    "    layer2 <- vlnplot$layers[[2]]\n",
    "    \n",
    "    vlnplot$layers[[1]] <- layer2\n",
    "    vlnplot$layers[[2]] <- layer1\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_vlnplot.pdf'), width = 5.5, height = 12)\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_vlnplot.eps'), width = 5.5, height = 12)}\n",
    "    \n",
    "    #create heatmap for the positive genes\n",
    "    \n",
    "    #create heatmap for positive genes\n",
    "    if(nrow(de_positive) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_positive <- object_av_wf@assays$RNA@scale.data[de_positive$genes, ] \n",
    "    ordered_genes_positive <- as.data.frame(ordered_genes_positive)\n",
    "    ordered_genes_positive <- ordered_genes_positive[order(-ordered_genes_positive[, cluster_of_interest[cl]]), ]\n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_positive <- rownames(ordered_genes_positive)} else {ordered_genes_positive <- rownames(de_positive)} \n",
    "    \n",
    "    #top 20 sorted\n",
    "    if(length(ordered_genes_positive) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_positive[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_positive}\n",
    "    #DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7,  angle = 270, hjust = 1,\n",
    "    heatmap <- DoHeatmap(object_av_wf, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups_wf) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap_without_false.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap_without_false.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #create heatmap for the negative genes\n",
    "    \n",
    "    if(nrow(de_negative) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_negative <- object_av@assays$RNA@scale.data[de_negative$genes, ] \n",
    "    ordered_genes_negative <- as.data.frame(ordered_genes_negative)\n",
    "    ordered_genes_negative <- ordered_genes_negative[order(-ordered_genes_negative[, group2]), ]\n",
    "    \n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_negative <- rownames(ordered_genes_negative)} else {ordered_genes_negative <- rownames(de_negative)}\n",
    "    \n",
    "    #top 30 sorted\n",
    "    if(length(ordered_genes_negative) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_negative[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_negative}\n",
    "    heatmap <- DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #build vlnplot\n",
    "    if(length(ordered_genes_plot) > 1){\n",
    "    vlnplot <- VlnPlot(object, features = ordered_genes_plot, stack = TRUE, flip = TRUE, cols = colours_diagnosis_groups, fill.by = \"ident\") +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))+ \n",
    "               geom_point_rast(position = 'jitter', size = 0.01, alpha = 0.7, colour = 'grey')\n",
    "    layer1 <- vlnplot$layers[[1]]\n",
    "    layer2 <- vlnplot$layers[[2]]\n",
    "    \n",
    "    vlnplot$layers[[1]] <- layer2\n",
    "    vlnplot$layers[[2]] <- layer1\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', group2, '_vlnplot.pdf'), width = 5.5, height = 12)\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', group2, '_vlnplot.eps'), width = 5.5, height = 12)}\n",
    "    \n",
    "    #create heatmap for negative genes\n",
    "    \n",
    "    if(nrow(de_negative) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_negative <- object_av_wf@assays$RNA@scale.data[de_negative$genes, ] \n",
    "    ordered_genes_negative <- as.data.frame(ordered_genes_negative)\n",
    "    ordered_genes_negative <- ordered_genes_negative[order(-ordered_genes_negative[, group2]), ]\n",
    "    \n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_negative <- rownames(ordered_genes_negative)} else {ordered_genes_negative <- rownames(de_negative)}\n",
    "    \n",
    "    #top 30 sorted\n",
    "    if(length(ordered_genes_negative) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_negative[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_negative}\n",
    "    heatmap <- DoHeatmap(object_av_wf, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups_wf) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap_without_FALSE.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap_without_FALSE.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    \n",
    "    #start the enrichr PEA \n",
    "    for(db in 1:length(databases_list)){\n",
    "        enriched <- enrichr(de_positive$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the positive enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_pea_list.csv'))\n",
    "            \n",
    "        \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        if(nrow(enriched) > 0){\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_positive <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour1) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl], databases_list[db]))\n",
    "            } else {\n",
    "            plot_positive <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour1) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl], databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_positive, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_PEA.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "        }\n",
    "        \n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE, group.colors = colours_diagnosis_groups) + theme(text = element_text(size = 20, face = \"bold\")) + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_',  '_HEATMAP_genes', '.pdf'), width = 5, height = 11)\n",
    "        \n",
    "        }\n",
    "        \n",
    "        \n",
    "        #perform the enrichment on negative genes\n",
    "        enriched <- enrichr(de_negative$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the negative enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_pea_list.csv'))\n",
    "            \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        if(nrow(enriched) > 0){\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_negative <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour2) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(group2, databases_list[db]))\n",
    "            } else {\n",
    "            plot_negative <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour2) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(group2, databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_negative, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_PEA','.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            genes_to_plot\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            genes_to_plot\n",
    "        }\n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE, group.colors = colours_diagnosis_groups) + theme(text = element_text(size = 20, face = \"bold\")) + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_HEATMAP_genes', '.pdf'), width = 5, height = 11)\n",
    "        }\n",
    "        }\n",
    "    #end of the einrichr loop\n",
    "}\n",
    "#end of the clusters loop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SYTL4 vs JU4/JU3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create grouping dge\n",
    "\n",
    "pbmc@meta.data$h1_general_all <- 'FALSE'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_KIF2C')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_JU')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_4JU') | startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_3JU')] <- '3or4JU'\n",
    "pbmc@meta.data$h1_general_all[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_SYTL4')] <- 'SYTL4'\n",
    "unique(pbmc@meta.data$h1_general_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for heatmaps\n",
    "pbmc@meta.data$h1_general <- 'FALSE'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_KIF2C')] <- 'KIF2C'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_JU')] <- 'JU'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_4JU') | startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_3JU')] <- '3or4JU'\n",
    "pbmc@meta.data$h1_general[startsWith(x = pbmc@meta.data$s_clonotypes, 'H1_SYTL4')] <- 'SYTL4'\n",
    "unique(pbmc@meta.data$h1_general)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# presort the features according to the expression within the samples of the group of interest\n",
    "#define the group of analysis, cluster of interest, clustering column, patient column\n",
    "cluster_of_interest <- c('3or4JU')\n",
    "group2 <- 'SYTL4'\n",
    "object <- pbmc\n",
    "grouping <- 'h1_general_all'\n",
    "cutoff_expression <- 0.05 #for the DGE min.pct\n",
    "logfc_cutoff <- 0.02 #for the DGE analysis\n",
    "padj_cutoff <- 0.05 #for the genes coming into the GSEA analysis\n",
    "databases_list <- c('GO_Biological_Process_2021', 'Reactome_2016')\n",
    "features <- rownames(object)\n",
    "markers.remove <- grep(pattern = \"^TRAV|^TRBV|^TRGV|^TRDV|^RPL|^RPS|^MT-|^IGKV|^IGLV|^IGHV|^IGH|^IGKC|^TRBC\", x = rownames(object), value = TRUE)\n",
    "features <- features[!(features%in%markers.remove)]\n",
    "\n",
    "Idents(object) <- 'h1_general'\n",
    "levels(object) <- c('KIF2C', 'JU', 'SYTL4', '3or4JU', 'FALSE')\n",
    "object_av <- AverageExpression(object, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "\n",
    "#for the heatmap\n",
    "object_wf <- subset(object, idents = 'FALSE', invert = TRUE)\n",
    "levels(object_wf) <- c('KIF2C', 'JU', 'SYTL4', '3or4JU')\n",
    "object_av_wf <- AverageExpression(object_wf, assay = \"RNA\", return.seurat = TRUE, verbose = FALSE)\n",
    "colours_diagnosis_groups_wf <- c('#5cadfb', '#1d9d01', '#ff8a01', '#8800ff')\n",
    "\n",
    "\n",
    "colours_diagnosis_groups <- c('#5cadfb', '#1d9d01', '#ff8a01', '#8800ff', '#bebebe')\n",
    "colour1 <- '#8800ff'\n",
    "colour2 <- '#ff8a01'\n",
    "\n",
    "#here starts the loop with clusters of interest\n",
    "for(cl in 1:length(cluster_of_interest)){\n",
    "    \n",
    "    #create the dir with output\n",
    "    dir_path <- paste0('outs/', cluster_of_interest[cl], '_vs_', group2, '_all')\n",
    "    dir.create(dir_path)\n",
    "    \n",
    "    de_genes <- FindMarkers(object = object, features = features, only.pos = FALSE, group.by = grouping, ident.1 = cluster_of_interest[cl], ident.2 = group2, min.pct = cutoff_expression, logfc.threshold = logfc_cutoff)\n",
    "    de_genes$genes <- rownames(de_genes)\n",
    "    ##prepare the plotting table (split by the log2FC and position the most negative in reverse)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #subset negative genes\n",
    "    de_negative <- de_genes[de_genes$avg_log2FC < 0, ]\n",
    "    de_negative <- de_negative[order(de_negative$avg_log2FC), ]\n",
    "    de_genes <- rbind(de_positive, de_negative)\n",
    "    \n",
    "    #build and save the volcano plot\n",
    "    volcano <- ggplot(de_genes, aes(x = avg_log2FC, y = -log10(p_val_adj))) +\n",
    "        geom_vline(xintercept = 0) +\n",
    "        geom_hline(yintercept = -log10(padj_cutoff), color =\"grey\", linetype =\"dashed\") +\n",
    "        geom_point(data = de_genes,\n",
    "                    color = \"grey\", alpha = 0.5) +\n",
    "        geom_point(data = subset(de_genes, avg_log2FC > 0 & p_val_adj < padj_cutoff)[1:20, ],\n",
    "                    fill = colour1, alpha = 1, shape=21, size= 2.5) +\n",
    "        geom_point(data = subset(de_genes, avg_log2FC < 0 & p_val_adj < padj_cutoff)[1:20, ],\n",
    "                    fill = colour2, alpha = 1, shape=21, size= 2.5) +\n",
    "        geom_text_repel(data=rbind(subset(de_genes, avg_log2FC > 0 & p_val_adj < padj_cutoff)[1:20, ], subset(de_genes, avg_log2FC < 0 & p_val_adj < padj_cutoff)[1:20, ]), max.overlaps = 50, aes(label = genes))+\n",
    "        theme_linedraw() +\n",
    "        theme(panel.grid = element_blank(), legend.position = \"none\", \n",
    "              plot.title = element_text(size = 15, hjust = 0.5)) + ggtitle(paste0(group2, ' (left)', ' vs ', cluster_of_interest, ' (right)')) + \n",
    "        xlab(\"log2(average fold change)\") +\n",
    "        ylab(\"-log10(p-value)\")\n",
    "    ggsave(volcano, file = paste0(dir_path, '/', cluster_of_interest[cl], '_',  '_vs_', group2, '_volcano_plot.pdf'), height = 8, width = 10)\n",
    "    \n",
    "    #prepare the subset of genes for PEA (filter for p value)\n",
    "    #subset positive genes\n",
    "    de_positive <- de_genes[de_genes$avg_log2FC > 0, ]\n",
    "    de_positive <- de_positive[de_positive$p_val_adj < padj_cutoff, ]\n",
    "    de_positive <- de_positive[order(-de_positive$avg_log2FC), ]\n",
    "    \n",
    "    #subset negative genes\n",
    "    de_negative <- de_genes[de_genes$avg_log2FC < 0, ]\n",
    "    de_negative <- de_negative[de_negative$p_val_adj < padj_cutoff, ]\n",
    "    de_negative <- de_negative[order(de_negative$avg_log2FC), ]\n",
    "    \n",
    "    de_all <- rbind(de_positive, de_negative)\n",
    "    #save the de_all \n",
    "    write.csv(de_all, file = paste0(dir_path, '/', cluster_of_interest[cl], '_',  '_vs_', group2, '_de_list.csv'))\n",
    "    \n",
    "    #create heatmap for positive genes\n",
    "    if(nrow(de_positive) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_positive <- object_av@assays$RNA@scale.data[de_positive$genes, ] \n",
    "    ordered_genes_positive <- as.data.frame(ordered_genes_positive)\n",
    "    ordered_genes_positive <- ordered_genes_positive[order(-ordered_genes_positive[, cluster_of_interest[cl]]), ]\n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_positive <- rownames(ordered_genes_positive)} else {ordered_genes_positive <- rownames(de_positive)} \n",
    "    \n",
    "    #top 20 sorted\n",
    "    if(length(ordered_genes_positive) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_positive[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_positive}\n",
    "    #DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7,  angle = 270, hjust = 1,\n",
    "    heatmap <- DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #build vlnplot\n",
    "    if(length(ordered_genes_plot) > 1){\n",
    "    vlnplot <- VlnPlot(object, features = ordered_genes_plot, stack = TRUE, flip = TRUE, cols = colours_diagnosis_groups, fill.by = \"ident\") +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))+ \n",
    "               geom_point_rast(position = 'jitter', size = 0.01, alpha = 0.7, colour = 'grey')\n",
    "    layer1 <- vlnplot$layers[[1]]\n",
    "    layer2 <- vlnplot$layers[[2]]\n",
    "    \n",
    "    vlnplot$layers[[1]] <- layer2\n",
    "    vlnplot$layers[[2]] <- layer1\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_vlnplot.pdf'), width = 5.5, height = 12)\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_vlnplot.eps'), width = 5.5, height = 12)}\n",
    "    \n",
    "    #create heatmap for the positive genes\n",
    "    \n",
    "    #create heatmap for positive genes\n",
    "    if(nrow(de_positive) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_positive <- object_av_wf@assays$RNA@scale.data[de_positive$genes, ] \n",
    "    ordered_genes_positive <- as.data.frame(ordered_genes_positive)\n",
    "    ordered_genes_positive <- ordered_genes_positive[order(-ordered_genes_positive[, cluster_of_interest[cl]]), ]\n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_positive <- rownames(ordered_genes_positive)} else {ordered_genes_positive <- rownames(de_positive)} \n",
    "    \n",
    "    #top 20 sorted\n",
    "    if(length(ordered_genes_positive) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_positive[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_positive}\n",
    "    #DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7,  angle = 270, hjust = 1,\n",
    "    heatmap <- DoHeatmap(object_av_wf, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups_wf) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap_without_false.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', cluster_of_interest[cl], '_heatmap_without_false.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #create heatmap for the negative genes\n",
    "    \n",
    "    if(nrow(de_negative) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_negative <- object_av@assays$RNA@scale.data[de_negative$genes, ] \n",
    "    ordered_genes_negative <- as.data.frame(ordered_genes_negative)\n",
    "    ordered_genes_negative <- ordered_genes_negative[order(-ordered_genes_negative[, group2]), ]\n",
    "    \n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_negative <- rownames(ordered_genes_negative)} else {ordered_genes_negative <- rownames(de_negative)}\n",
    "    \n",
    "    #top 30 sorted\n",
    "    if(length(ordered_genes_negative) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_negative[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_negative}\n",
    "    heatmap <- DoHeatmap(object_av, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    #build vlnplot\n",
    "    if(length(ordered_genes_plot) > 1){\n",
    "    vlnplot <- VlnPlot(object, features = ordered_genes_plot, stack = TRUE, flip = TRUE, cols = colours_diagnosis_groups, fill.by = \"ident\") +\n",
    "                        theme(legend.position = \"none\",\n",
    "                          text = element_text(size = 17),\n",
    "                          axis.text = element_text(size = 17))+ \n",
    "               geom_point_rast(position = 'jitter', size = 0.01, alpha = 0.7, colour = 'grey')\n",
    "    layer1 <- vlnplot$layers[[1]]\n",
    "    layer2 <- vlnplot$layers[[2]]\n",
    "    \n",
    "    vlnplot$layers[[1]] <- layer2\n",
    "    vlnplot$layers[[2]] <- layer1\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', group2, '_vlnplot.pdf'), width = 5.5, height = 12)\n",
    "    ggsave(vlnplot, file = paste0(dir_path, '/', group2, '_vlnplot.eps'), width = 5.5, height = 12)}\n",
    "    \n",
    "    #create heatmap for negative genes\n",
    "    \n",
    "    if(nrow(de_negative) > 1){\n",
    "    #create ordered heatmap\n",
    "    ordered_genes_negative <- object_av_wf@assays$RNA@scale.data[de_negative$genes, ] \n",
    "    ordered_genes_negative <- as.data.frame(ordered_genes_negative)\n",
    "    ordered_genes_negative <- ordered_genes_negative[order(-ordered_genes_negative[, group2]), ]\n",
    "    \n",
    "    #prepare the genes for heatmap\n",
    "    ordered_genes_negative <- rownames(ordered_genes_negative)} else {ordered_genes_negative <- rownames(de_negative)}\n",
    "    \n",
    "    #top 30 sorted\n",
    "    if(length(ordered_genes_negative) > 30){\n",
    "    ordered_genes_plot <- ordered_genes_negative[1:30]\n",
    "    } else {ordered_genes_plot <- ordered_genes_negative}\n",
    "    heatmap <- DoHeatmap(object_av_wf, features = ordered_genes_plot, draw.lines = FALSE, size = 7, raster = FALSE, group.colors = colours_diagnosis_groups_wf) + \n",
    "    theme(\n",
    "        text = element_text(size = 19, colour = 'black', face = 'plain'),\n",
    "        axis.text.y = element_text(size = 19, colour = 'black', face = 'plain')) +\n",
    "    scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap_without_FALSE.pdf'), width = 5.5, height = 11)\n",
    "    ggsave(heatmap, file = paste0(dir_path, '/', group2, '_heatmap_without_FALSE.eps'), width = 5.5, height = 11)\n",
    "    \n",
    "    \n",
    "    #start the enrichr PEA \n",
    "    for(db in 1:length(databases_list)){\n",
    "        enriched <- enrichr(de_positive$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the positive enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_pea_list.csv'))\n",
    "            \n",
    "        \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        if(nrow(enriched) > 0){\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_positive <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour1) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl], databases_list[db]))\n",
    "            } else {\n",
    "            plot_positive <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour1) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(cluster_of_interest[cl], databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_positive, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_PEA.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "        }\n",
    "        \n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE, group.colors = colours_diagnosis_groups) + theme(text = element_text(size = 20, face = \"bold\")) + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', cluster_of_interest[cl], '_', databases_list[db], '_',  '_HEATMAP_genes', '.pdf'), width = 5, height = 11)\n",
    "        \n",
    "        }\n",
    "        \n",
    "        \n",
    "        #perform the enrichment on negative genes\n",
    "        enriched <- enrichr(de_negative$gene, databases = databases_list[db])\n",
    "        enriched <- enriched[[1]]\n",
    "        enriched <- enriched[order(-enriched$Adjusted.P.value), ]\n",
    "        enriched$Term <- factor(enriched$Term, levels = unique(enriched$Term))\n",
    "        #save the negative enriched pathways\n",
    "        write.csv(enriched, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_pea_list.csv'))\n",
    "            \n",
    "        reverselog_trans <- function(base = exp(1)) {\n",
    "            trans <- function(x) -log(x, base)\n",
    "            inv <- function(x) base^(-x)\n",
    "            trans_new(paste0(\"reverselog-\", format(base)), trans, inv,\n",
    "                      log_breaks(base = base),\n",
    "                      domain = c(1e-100, Inf))\n",
    "            }\n",
    "        #2881c1 - for blue\n",
    "        #d3556e - for red\n",
    "        if(nrow(enriched) > 0){\n",
    "        options(repr.plot.width=22, repr.plot.height=11)\n",
    "        if(nrow(enriched) > 20){\n",
    "            plot_negative <- ggplot(enriched[(nrow(enriched)-19):nrow(enriched), ], aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour2) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(group2, databases_list[db]))\n",
    "            } else {\n",
    "            plot_negative <- ggplot(enriched, aes(y=Term, x= Adjusted.P.value))+\n",
    "                geom_vline(xintercept = .05, color = \"grey\", linetype=\"dashed\") +\n",
    "                geom_segment( aes(yend=Term, xend=1), col= \"black\") +\n",
    "                geom_point(shape=21, aes(size = abs(Combined.Score)), fill = colour2) +\n",
    "                scale_x_continuous(trans=reverselog_trans(10))+\n",
    "                scale_size_continuous(range = c(2, 13)) +\n",
    "                theme_tufte()+ xlab(\"p value (adj)\") + ylab(\"\") +\n",
    "                theme(text=element_text(family=\"Helvetica\", size = 26),\n",
    "                  axis.text.x = element_text(angle = 45, hjust = 1), legend.position = \"right\") + ggtitle(paste(group2, databases_list[db]))\n",
    "        }\n",
    "        ggsave(plot_negative, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_PEA','.pdf'), width = 30, height = 11)\n",
    "            \n",
    "        #plot the genes\n",
    "        if(nrow(enriched) > 20){ \n",
    "            genes_to_plot <- paste0(x = enriched[(nrow(enriched)-19):nrow(enriched), 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            #reverse because of ascending ordering of the enriched table to put the most significant at the beginning\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            genes_to_plot\n",
    "            }else{\n",
    "            genes_to_plot <- paste0(x = enriched[, 'Genes'], ';')\n",
    "            genes_to_plot <- paste(genes_to_plot, collapse = '')\n",
    "            genes_to_plot <- strsplit(genes_to_plot, split = ';')[[1]]\n",
    "            genes_to_plot <- rev(genes_to_plot)\n",
    "            genes_to_plot <- unique(genes_to_plot)\n",
    "            genes_to_plot\n",
    "        }\n",
    "        if(length(genes_to_plot) > 20){genes_to_plot <- genes_to_plot[1:20]}\n",
    "        heatmap_plot <- DoHeatmap(object_av, features = genes_to_plot, draw.lines = FALSE, raster = FALSE, group.colors = colours_diagnosis_groups) + theme(text = element_text(size = 20, face = \"bold\")) + scale_fill_gradientn(colors = c(\"#2881C1\", \"white\", \"#D3556E\"))\n",
    "        ggsave(heatmap_plot, file = paste0(dir_path, '/', group2, '_', databases_list[db], '_HEATMAP_genes', '.pdf'), width = 5, height = 11)\n",
    "        }\n",
    "        }\n",
    "    #end of the einrichr loop\n",
    "}\n",
    "#end of the clusters loop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save h5ad file for the diffusion maps analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "levels(pbmc)\n",
    "object <- subset(pbmc, idents = c('MAIT_1', 'MAIT_2'), invert = TRUE)\n",
    "DefaultAssay(object) <- 'integrated'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "levels(object)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library('SeuratData')\n",
    "library('SeuratDisk')\n",
    "SaveH5Seurat(object, filename = \"pbmc_subset_fordiffusion.h5Seurat\")\n",
    "Convert(\"pbmc_subset_fordiffusion.h5Seurat\", dest = \"h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Diffusion map: in Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as pl\n",
    "from matplotlib import rcParams\n",
    "import scanpy as sc\n",
    "import scipy as sp\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import umap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.settings.set_figure_params(dpi=150, frameon=False, figsize=(7, 7), facecolor='white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from platform import python_version\n",
    "\n",
    "print(python_version())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Generate diffusion maps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#load the object\n",
    "random.seed(0)\n",
    "adata = sc.read_h5ad(\"./pbmc_subset_fordiffusion.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.highly_variable_genes(adata, n_top_genes = 1000, flavor=\"seurat\")\n",
    "sc.tl.pca(adata)\n",
    "#adata.obs[\"cd8names\"] = adata.obs[\"cd8names\"].apply(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata, n_neighbors=20, n_pcs=18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.diffmap(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "palette = [\"#a6cee3\", \"#1e78b4\", \"#b2df8a\", \"#34a12d\", \"#fa9a99\", \"#e31a1c\", \"#fcbf6f\", \"#ff7f01\", \"#cab2d6\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "sc.set_figure_params(dpi=300, frameon=False, figsize=(10, 3), facecolor='white')\n",
    "sc.pl.diffmap(adata,components=[1,2],color=[\"cd8names\"], frameon=True, projection = \"2d\",\n",
    "              legend_loc=\"right margin\", palette = palette, legend_fontsize = 12, legend_fontweight = 'normal', size = 20,\n",
    "             save = '_withlegend.pdf')\n",
    "\n",
    "sc.set_figure_params(dpi=300, frameon=False, figsize=(9, 3), facecolor='white')\n",
    "sc.pl.diffmap(adata,components=[1,2],color=[\"cd8names\"], frameon=True, projection = \"2d\",\n",
    "              legend_loc=\"on data\", palette = palette, legend_fontsize = 7, legend_fontweight = 'normal', size = 20,\n",
    "             save = '.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add the pseudotime score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#annotate the root and compute pseudotime\n",
    "adata.uns['iroot'] = np.flatnonzero(adata.obs['cd8names'] == '1_CCR7')[0]\n",
    "sc.tl.dpt(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#visualize the new pseudotime\n",
    "sc.set_figure_params(dpi=300, frameon=False, figsize=(10, 3), facecolor='white', vector_friendly = True)\n",
    "sc.pl.diffmap(adata,components=[1,2],color=[\"dpt_pseudotime\"], frameon=True, projection = \"2d\",\n",
    "              legend_loc=\"on data\", color_map = 'binary', legend_fontsize = 12, legend_fontweight = 'normal', size = 20, sort_order = True, save = '_withdm_pseudotime_bw.pdf')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.0.5 azimuth env",
   "language": "R",
   "name": "ir405"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
